// @cloudscape-design/components/internal/generated/styles/tokens.js
export var colorPrimary50 = "var(--color-primary-50-1y05xv, #f0fbff)";
export var colorPrimary100 = "var(--color-primary-100-f62fz9, #d1f1ff)";
export var colorPrimary200 = "var(--color-primary-200-vubr4w, #b8e7ff)";
export var colorPrimary300 = "var(--color-primary-300-5q65ox, #75cfff)";
export var colorPrimary400 = "var(--color-primary-400-n8h4bx, #42b4ff)";
export var colorPrimary500 = "var(--color-primary-500-q9c16y, #0099ff)";
export var colorPrimary600 = "var(--color-primary-600-1lcy1k, #006ce0)";
export var colorPrimary700 = "var(--color-primary-700-n6k121, #004a9e)";
export var colorPrimary800 = "var(--color-primary-800-j9rj38, #003b8f)";
export var colorPrimary900 = "var(--color-primary-900-a5kqrr, #002b66)";
export var colorPrimary1000 = "var(--color-primary-1000-7umopx, #001129)";
export var colorNeutral50 = "var(--color-neutral-50-pvu04n, #fcfcfd)";
export var colorNeutral100 = "var(--color-neutral-100-gk3lvf, #f9f9fa)";
export var colorNeutral150 = "var(--color-neutral-150-gezhen, #f6f6f9)";
export var colorNeutral200 = "var(--color-neutral-200-fqt4tz, #f3f3f7)";
export var colorNeutral250 = "var(--color-neutral-250-vs1is4, #ebebf0)";
export var colorNeutral300 = "var(--color-neutral-300-08wi6k, #dedee3)";
export var colorNeutral350 = "var(--color-neutral-350-dq6kfr, #c6c6cd)";
export var colorNeutral400 = "var(--color-neutral-400-wtst55, #b4b4bb)";
export var colorNeutral450 = "var(--color-neutral-450-kn0235, #a4a4ad)";
export var colorNeutral500 = "var(--color-neutral-500-8van0b, #8c8c94)";
export var colorNeutral550 = "var(--color-neutral-550-z2a44u, #72747e)";
export var colorNeutral600 = "var(--color-neutral-600-fln1ww, #656871)";
export var colorNeutral650 = "var(--color-neutral-650-miik4f, #424650)";
export var colorNeutral700 = "var(--color-neutral-700-qw8ats, #333843)";
export var colorNeutral750 = "var(--color-neutral-750-pi9qqd, #232b37)";
export var colorNeutral800 = "var(--color-neutral-800-t7j5ap, #1b232d)";
export var colorNeutral850 = "var(--color-neutral-850-3f0gro, #161d26)";
export var colorNeutral900 = "var(--color-neutral-900-v0mtoc, #131920)";
export var colorNeutral950 = "var(--color-neutral-950-lxybh8, #0f141a)";
export var colorNeutral1000 = "var(--color-neutral-1000-7ovvlt, #06080a)";
export var colorError50 = "var(--color-error-50-1upkvz, #fff5f5)";
export var colorError400 = "var(--color-error-400-c0knb8, #ff7a7a)";
export var colorError600 = "var(--color-error-600-mdn3ng, #db0000)";
export var colorError900 = "var(--color-error-900-skonp1, #700000)";
export var colorError1000 = "var(--color-error-1000-ecmudm, #1f0000)";
export var colorSuccess50 = "var(--color-success-50-yow9uc, #effff1)";
export var colorSuccess500 = "var(--color-success-500-hhxb4g, #2bb534)";
export var colorSuccess600 = "var(--color-success-600-g7hz2i, #00802f)";
export var colorSuccess1000 = "var(--color-success-1000-xy1gvq, #001401)";
export var colorWarning50 = "var(--color-warning-50-uxheb8, #fffef0)";
export var colorWarning400 = "var(--color-warning-400-55puga, #ffe347)";
export var colorWarning500 = "var(--color-warning-500-zidhub, #fbd332)";
export var colorWarning900 = "var(--color-warning-900-kpxt8c, #855900)";
export var colorWarning1000 = "var(--color-warning-1000-65l070, #191100)";
export var colorInfo50 = "var(--color-info-50-cdvtrs, #f0fbff)";
export var colorInfo300 = "var(--color-info-300-q9xd1l, #75cfff)";
export var colorInfo400 = "var(--color-info-400-674xac, #42b4ff)";
export var colorInfo600 = "var(--color-info-600-un21zh, #006ce0)";
export var colorInfo1000 = "var(--color-info-1000-s77ok9, #001129)";
export var colorGrey50 = "var(--color-grey-50-mhbfod, #fcfcfd)";
export var colorGrey100 = "var(--color-grey-100-2sczed, #f9f9fa)";
export var colorGrey150 = "var(--color-grey-150-znah9n, #f6f6f9)";
export var colorGrey200 = "var(--color-grey-200-s986pt, #f3f3f7)";
export var colorGrey250 = "var(--color-grey-250-l5wm60, #ebebf0)";
export var colorGrey300 = "var(--color-grey-300-tqmb21, #dedee3)";
export var colorGrey350 = "var(--color-grey-350-brqf9i, #c6c6cd)";
export var colorGrey400 = "var(--color-grey-400-oueki4, #b4b4bb)";
export var colorGrey450 = "var(--color-grey-450-6n5usg, #a4a4ad)";
export var colorGrey500 = "var(--color-grey-500-argwpz, #8c8c94)";
export var colorGrey600 = "var(--color-grey-600-qux9qr, #656871)";
export var colorGrey650 = "var(--color-grey-650-5vpn76, #424650)";
export var colorGrey700 = "var(--color-grey-700-i9hww0, #333843)";
export var colorGrey750 = "var(--color-grey-750-3a32f4, #232b37)";
export var colorGrey800 = "var(--color-grey-800-38aulk, #1b232d)";
export var colorGrey850 = "var(--color-grey-850-fg4m0u, #161d26)";
export var colorGrey900 = "var(--color-grey-900-2zzggd, #131920)";
export var colorGrey950 = "var(--color-grey-950-rpj9g8, #0f141a)";
export var colorGrey1000 = "var(--color-grey-1000-kdxctl, #06080a)";
export var colorBlue50 = "var(--color-blue-50-1qtweg, #f0fbff)";
export var colorBlue100 = "var(--color-blue-100-a7atso, #d1f1ff)";
export var colorBlue200 = "var(--color-blue-200-66lcr3, #b8e7ff)";
export var colorBlue300 = "var(--color-blue-300-s2r2o0, #75cfff)";
export var colorBlue400 = "var(--color-blue-400-h2lqfk, #42b4ff)";
export var colorBlue600 = "var(--color-blue-600-dkpfum, #006ce0)";
export var colorBlue700 = "var(--color-blue-700-lcgdon, #004a9e)";
export var colorBlue900 = "var(--color-blue-900-k0czor, #002b66)";
export var colorBlue1000 = "var(--color-blue-1000-fmz0jt, #001129)";
export var colorGreen50 = "var(--color-green-50-p1p8l4, #effff1)";
export var colorGreen500 = "var(--color-green-500-c2uzi6, #2bb534)";
export var colorGreen600 = "var(--color-green-600-vdmico, #00802f)";
export var colorGreen900 = "var(--color-green-900-6aqp28, #00471e)";
export var colorGreen1000 = "var(--color-green-1000-tarlm7, #001401)";
export var colorRed50 = "var(--color-red-50-k823mj, #fff5f5)";
export var colorRed400 = "var(--color-red-400-40ay29, #ff7a7a)";
export var colorRed600 = "var(--color-red-600-lxm4n7, #db0000)";
export var colorRed900 = "var(--color-red-900-gj885t, #700000)";
export var colorRed1000 = "var(--color-red-1000-jtashw, #1f0000)";
export var colorYellow50 = "var(--color-yellow-50-wk1vy0, #fffef0)";
export var colorYellow400 = "var(--color-yellow-400-q4m2pa, #ffe347)";
export var colorYellow500 = "var(--color-yellow-500-rj98kw, #fbd332)";
export var colorYellow900 = "var(--color-yellow-900-jipa9u, #855900)";
export var colorYellow1000 = "var(--color-yellow-1000-zo0nn2, #191100)";
export var colorPurple400 = "var(--color-purple-400-o7fr4k, #bf80ff)";
export var colorPurple700 = "var(--color-purple-700-5hcbeu, #7300e5)";
export var colorAmber400 = "var(--color-amber-400-z6oddn, #ff9900)";
export var colorAmber500 = "var(--color-amber-500-rsho1x, #fa6f00)";
export var colorAwsSquidInk = "var(--color-aws-squid-ink-oxiega, #232f3e)";
export var colorTransparent = "var(--color-transparent-i61gs1, transparent)";
export var colorBlack = "var(--color-black-cox1hy, #000000)";
export var colorWhite = "var(--color-white-p1zlvy, #ffffff)";
export var colorChartsRed300 = "var(--color-charts-red-300-2k7eul, #ea7158)";
export var colorChartsRed400 = "var(--color-charts-red-400-ssrf2o, #dc5032)";
export var colorChartsRed500 = "var(--color-charts-red-500-m14kmu, #d13313)";
export var colorChartsRed600 = "var(--color-charts-red-600-938v3h, #ba2e0f)";
export var colorChartsRed700 = "var(--color-charts-red-700-f6sq8t, #a82a0c)";
export var colorChartsRed800 = "var(--color-charts-red-800-tzkaad, #972709)";
export var colorChartsRed900 = "var(--color-charts-red-900-fhg0lh, #892407)";
export var colorChartsRed1000 = "var(--color-charts-red-1000-9iigzo, #7d2105)";
export var colorChartsRed1100 = "var(--color-charts-red-1100-4n7b3z, #721e03)";
export var colorChartsRed1200 = "var(--color-charts-red-1200-ek3cuo, #671c00)";
export var colorChartsOrange300 = "var(--color-charts-orange-300-hqhtmn, #e07941)";
export var colorChartsOrange400 = "var(--color-charts-orange-400-g8c1fc, #cc5f21)";
export var colorChartsOrange500 = "var(--color-charts-orange-500-j3c2cu, #bc4d01)";
export var colorChartsOrange600 = "var(--color-charts-orange-600-1ad7o4, #a84401)";
export var colorChartsOrange700 = "var(--color-charts-orange-700-spsf2r, #983c02)";
export var colorChartsOrange800 = "var(--color-charts-orange-800-244d7b, #8a3603)";
export var colorChartsOrange900 = "var(--color-charts-orange-900-8omk92, #7e3103)";
export var colorChartsOrange1000 = "var(--color-charts-orange-1000-ezq5pz, #732c02)";
export var colorChartsOrange1100 = "var(--color-charts-orange-1100-bhcmg5, #692801)";
export var colorChartsOrange1200 = "var(--color-charts-orange-1200-exs6jj, #602400)";
export var colorChartsYellow300 = "var(--color-charts-yellow-300-fpz8o0, #b2911c)";
export var colorChartsYellow400 = "var(--color-charts-yellow-400-vxiqrf, #9c7b0b)";
export var colorChartsYellow500 = "var(--color-charts-yellow-500-1qgrtj, #8a6b05)";
export var colorChartsYellow600 = "var(--color-charts-yellow-600-aweqy2, #7b5f04)";
export var colorChartsYellow700 = "var(--color-charts-yellow-700-xh0lj9, #6f5504)";
export var colorChartsYellow800 = "var(--color-charts-yellow-800-gp4422, #654d03)";
export var colorChartsYellow900 = "var(--color-charts-yellow-900-r6gx3k, #5d4503)";
export var colorChartsYellow1000 = "var(--color-charts-yellow-1000-6dnac6, #553f03)";
export var colorChartsYellow1100 = "var(--color-charts-yellow-1100-fqp4sw, #4d3901)";
export var colorChartsYellow1200 = "var(--color-charts-yellow-1200-k7kf4w, #483300)";
export var colorChartsGreen300 = "var(--color-charts-green-300-6766ev, #67a353)";
export var colorChartsGreen400 = "var(--color-charts-green-400-gd41ay, #41902c)";
export var colorChartsGreen500 = "var(--color-charts-green-500-yr18n3, #1f8104)";
export var colorChartsGreen600 = "var(--color-charts-green-600-b1gmr0, #1a7302)";
export var colorChartsGreen700 = "var(--color-charts-green-700-305sle, #176702)";
export var colorChartsGreen800 = "var(--color-charts-green-800-rh42zr, #145d02)";
export var colorChartsGreen900 = "var(--color-charts-green-900-2x5smm, #125502)";
export var colorChartsGreen1000 = "var(--color-charts-green-1000-opphoq, #104d01)";
export var colorChartsGreen1100 = "var(--color-charts-green-1100-o67uzm, #0f4601)";
export var colorChartsGreen1200 = "var(--color-charts-green-1200-sw46fc, #0d4000)";
export var colorChartsTeal300 = "var(--color-charts-teal-300-2qlyrg, #2ea597)";
export var colorChartsTeal400 = "var(--color-charts-teal-400-s8pa77, #1c8e81)";
export var colorChartsTeal500 = "var(--color-charts-teal-500-8d830b, #0d7d70)";
export var colorChartsTeal600 = "var(--color-charts-teal-600-772n9t, #096f64)";
export var colorChartsTeal700 = "var(--color-charts-teal-700-29mnwm, #06645a)";
export var colorChartsTeal800 = "var(--color-charts-teal-800-vp41t6, #045b52)";
export var colorChartsTeal900 = "var(--color-charts-teal-900-k020ya, #03524a)";
export var colorChartsTeal1000 = "var(--color-charts-teal-1000-fhpqt3, #014b44)";
export var colorChartsTeal1100 = "var(--color-charts-teal-1100-6w598w, #01443e)";
export var colorChartsTeal1200 = "var(--color-charts-teal-1200-0u78my, #003e38)";
export var colorChartsBlue1300 = "var(--color-charts-blue-1-300-pdza0q, #529ccb)";
export var colorChartsBlue1400 = "var(--color-charts-blue-1-400-ajl038, #3184c2)";
export var colorChartsBlue1500 = "var(--color-charts-blue-1-500-9s8gor, #0273bb)";
export var colorChartsBlue1600 = "var(--color-charts-blue-1-600-7ymb7g, #0166ab)";
export var colorChartsBlue1700 = "var(--color-charts-blue-1-700-5qzras, #015b9d)";
export var colorChartsBlue1800 = "var(--color-charts-blue-1-800-awczh4, #015292)";
export var colorChartsBlue1900 = "var(--color-charts-blue-1-900-6wxwzk, #014a87)";
export var colorChartsBlue11000 = "var(--color-charts-blue-1-1000-00005b, #01437d)";
export var colorChartsBlue11100 = "var(--color-charts-blue-1-1100-8nwfwf, #003c75)";
export var colorChartsBlue11200 = "var(--color-charts-blue-1-1200-v60p8b, #00366d)";
export var colorChartsBlue2300 = "var(--color-charts-blue-2-300-g72slq, #688ae8)";
export var colorChartsBlue2400 = "var(--color-charts-blue-2-400-he538m, #5978e3)";
export var colorChartsBlue2500 = "var(--color-charts-blue-2-500-quctxu, #4066df)";
export var colorChartsBlue2600 = "var(--color-charts-blue-2-600-6qav3j, #3759ce)";
export var colorChartsBlue2700 = "var(--color-charts-blue-2-700-sp7t4m, #314fbf)";
export var colorChartsBlue2800 = "var(--color-charts-blue-2-800-q01umt, #2c46b1)";
export var colorChartsBlue2900 = "var(--color-charts-blue-2-900-gog7z2, #273ea5)";
export var colorChartsBlue21000 = "var(--color-charts-blue-2-1000-c13nf8, #23379b)";
export var colorChartsBlue21100 = "var(--color-charts-blue-2-1100-ddk6eo, #1f3191)";
export var colorChartsBlue21200 = "var(--color-charts-blue-2-1200-gt550t, #1b2b88)";
export var colorChartsPurple300 = "var(--color-charts-purple-300-85q036, #a783e1)";
export var colorChartsPurple400 = "var(--color-charts-purple-400-9axh6r, #9469d6)";
export var colorChartsPurple500 = "var(--color-charts-purple-500-rn2jbl, #8456ce)";
export var colorChartsPurple600 = "var(--color-charts-purple-600-26s4rg, #7749bf)";
export var colorChartsPurple700 = "var(--color-charts-purple-700-tv8cvg, #6b40b2)";
export var colorChartsPurple800 = "var(--color-charts-purple-800-h61qlx, #6237a7)";
export var colorChartsPurple900 = "var(--color-charts-purple-900-am452b, #59309d)";
export var colorChartsPurple1000 = "var(--color-charts-purple-1000-uarqpb, #512994)";
export var colorChartsPurple1100 = "var(--color-charts-purple-1100-y8ctnd, #4a238b)";
export var colorChartsPurple1200 = "var(--color-charts-purple-1200-hr9f40, #431d84)";
export var colorChartsPink300 = "var(--color-charts-pink-300-ewnht7, #da7596)";
export var colorChartsPink400 = "var(--color-charts-pink-400-smjdat, #ce567c)";
export var colorChartsPink500 = "var(--color-charts-pink-500-bw864b, #c33d69)";
export var colorChartsPink600 = "var(--color-charts-pink-600-2ro14y, #b1325c)";
export var colorChartsPink700 = "var(--color-charts-pink-700-ryxvua, #a32952)";
export var colorChartsPink800 = "var(--color-charts-pink-800-tcusf8, #962249)";
export var colorChartsPink900 = "var(--color-charts-pink-900-kpyne4, #8b1b42)";
export var colorChartsPink1000 = "var(--color-charts-pink-1000-1soluc, #81143b)";
export var colorChartsPink1100 = "var(--color-charts-pink-1100-ff6g93, #780d35)";
export var colorChartsPink1200 = "var(--color-charts-pink-1200-w9585d, #6f062f)";
export var colorChartsStatusCritical = "var(--color-charts-status-critical-c6brdu, #7d2105)";
export var colorChartsStatusHigh = "var(--color-charts-status-high-18fhg5, #ba2e0f)";
export var colorChartsStatusMedium = "var(--color-charts-status-medium-3trmy3, #cc5f21)";
export var colorChartsStatusLow = "var(--color-charts-status-low-br6wv0, #b2911c)";
export var colorChartsStatusPositive = "var(--color-charts-status-positive-md7eqa, #67a353)";
export var colorChartsStatusInfo = "var(--color-charts-status-info-yds4x2, #3184c2)";
export var colorChartsStatusNeutral = "var(--color-charts-status-neutral-k2p33t, #8c8c94)";
export var colorChartsThresholdNegative = "var(--color-charts-threshold-negative-aad26m, #db0000)";
export var colorChartsThresholdPositive = "var(--color-charts-threshold-positive-mk2804, #00802f)";
export var colorChartsThresholdInfo = "var(--color-charts-threshold-info-ijuzzj, #006ce0)";
export var colorChartsThresholdNeutral = "var(--color-charts-threshold-neutral-pd7kh4, #656871)";
export var colorChartsLineGrid = "var(--color-charts-line-grid-kjxf3m, #dedee3)";
export var colorChartsLineTick = "var(--color-charts-line-tick-xmcbvk, #dedee3)";
export var colorChartsLineAxis = "var(--color-charts-line-axis-b95ncf, #dedee3)";
export var colorChartsPaletteCategorical1 = "var(--color-charts-palette-categorical-1-xu0deg, #688ae8)";
export var colorChartsPaletteCategorical2 = "var(--color-charts-palette-categorical-2-ktit09, #c33d69)";
export var colorChartsPaletteCategorical3 = "var(--color-charts-palette-categorical-3-g0srj0, #2ea597)";
export var colorChartsPaletteCategorical4 = "var(--color-charts-palette-categorical-4-5vauwp, #8456ce)";
export var colorChartsPaletteCategorical5 = "var(--color-charts-palette-categorical-5-3v8ery, #e07941)";
export var colorChartsPaletteCategorical6 = "var(--color-charts-palette-categorical-6-ztdd8d, #3759ce)";
export var colorChartsPaletteCategorical7 = "var(--color-charts-palette-categorical-7-3j5o6w, #962249)";
export var colorChartsPaletteCategorical8 = "var(--color-charts-palette-categorical-8-c5r39m, #096f64)";
export var colorChartsPaletteCategorical9 = "var(--color-charts-palette-categorical-9-8n6iuv, #6237a7)";
export var colorChartsPaletteCategorical10 = "var(--color-charts-palette-categorical-10-opta0w, #a84401)";
export var colorChartsPaletteCategorical11 = "var(--color-charts-palette-categorical-11-b2r7jc, #273ea5)";
export var colorChartsPaletteCategorical12 = "var(--color-charts-palette-categorical-12-b5drtm, #780d35)";
export var colorChartsPaletteCategorical13 = "var(--color-charts-palette-categorical-13-c69xg9, #03524a)";
export var colorChartsPaletteCategorical14 = "var(--color-charts-palette-categorical-14-db19x8, #4a238b)";
export var colorChartsPaletteCategorical15 = "var(--color-charts-palette-categorical-15-8z8vjw, #7e3103)";
export var colorChartsPaletteCategorical16 = "var(--color-charts-palette-categorical-16-549jkl, #1b2b88)";
export var colorChartsPaletteCategorical17 = "var(--color-charts-palette-categorical-17-nrio7t, #ce567c)";
export var colorChartsPaletteCategorical18 = "var(--color-charts-palette-categorical-18-tm902v, #003e38)";
export var colorChartsPaletteCategorical19 = "var(--color-charts-palette-categorical-19-ujcr86, #9469d6)";
export var colorChartsPaletteCategorical20 = "var(--color-charts-palette-categorical-20-h55e4g, #602400)";
export var colorChartsPaletteCategorical21 = "var(--color-charts-palette-categorical-21-vs0u8l, #4066df)";
export var colorChartsPaletteCategorical22 = "var(--color-charts-palette-categorical-22-6klt3l, #a32952)";
export var colorChartsPaletteCategorical23 = "var(--color-charts-palette-categorical-23-3zpkdt, #0d7d70)";
export var colorChartsPaletteCategorical24 = "var(--color-charts-palette-categorical-24-z9a4uk, #6b40b2)";
export var colorChartsPaletteCategorical25 = "var(--color-charts-palette-categorical-25-tgdsk2, #bc4d01)";
export var colorChartsPaletteCategorical26 = "var(--color-charts-palette-categorical-26-lo8zn9, #2c46b1)";
export var colorChartsPaletteCategorical27 = "var(--color-charts-palette-categorical-27-bruhsa, #81143b)";
export var colorChartsPaletteCategorical28 = "var(--color-charts-palette-categorical-28-6b00fb, #045b52)";
export var colorChartsPaletteCategorical29 = "var(--color-charts-palette-categorical-29-aurmid, #512994)";
export var colorChartsPaletteCategorical30 = "var(--color-charts-palette-categorical-30-fjnmd7, #8a3603)";
export var colorChartsPaletteCategorical31 = "var(--color-charts-palette-categorical-31-7zcct5, #1f3191)";
export var colorChartsPaletteCategorical32 = "var(--color-charts-palette-categorical-32-rrda6y, #da7596)";
export var colorChartsPaletteCategorical33 = "var(--color-charts-palette-categorical-33-2v0mzv, #01443e)";
export var colorChartsPaletteCategorical34 = "var(--color-charts-palette-categorical-34-g9a9q3, #a783e1)";
export var colorChartsPaletteCategorical35 = "var(--color-charts-palette-categorical-35-u0w821, #692801)";
export var colorChartsPaletteCategorical36 = "var(--color-charts-palette-categorical-36-tthuf8, #5978e3)";
export var colorChartsPaletteCategorical37 = "var(--color-charts-palette-categorical-37-y588bl, #b1325c)";
export var colorChartsPaletteCategorical38 = "var(--color-charts-palette-categorical-38-qdh97u, #1c8e81)";
export var colorChartsPaletteCategorical39 = "var(--color-charts-palette-categorical-39-yisq6l, #7749bf)";
export var colorChartsPaletteCategorical40 = "var(--color-charts-palette-categorical-40-yeer1v, #cc5f21)";
export var colorChartsPaletteCategorical41 = "var(--color-charts-palette-categorical-41-tu9dxw, #314fbf)";
export var colorChartsPaletteCategorical42 = "var(--color-charts-palette-categorical-42-q410kp, #8b1b42)";
export var colorChartsPaletteCategorical43 = "var(--color-charts-palette-categorical-43-dwew7q, #06645a)";
export var colorChartsPaletteCategorical44 = "var(--color-charts-palette-categorical-44-2thp96, #59309d)";
export var colorChartsPaletteCategorical45 = "var(--color-charts-palette-categorical-45-6kinj6, #983c02)";
export var colorChartsPaletteCategorical46 = "var(--color-charts-palette-categorical-46-iefxfq, #23379b)";
export var colorChartsPaletteCategorical47 = "var(--color-charts-palette-categorical-47-9l9wl2, #6f062f)";
export var colorChartsPaletteCategorical48 = "var(--color-charts-palette-categorical-48-5s2n0r, #014b44)";
export var colorChartsPaletteCategorical49 = "var(--color-charts-palette-categorical-49-bub0l1, #431d84)";
export var colorChartsPaletteCategorical50 = "var(--color-charts-palette-categorical-50-utrpu5, #732c02)";
export var colorChartsErrorBarMarker = "var(--color-charts-error-bar-marker-r10jgv, #131920)";
export var colorSeverityDarkRed = "var(--color-severity-dark-red-j8bmoc, #870303)";
export var colorSeverityRed = "var(--color-severity-red-wf5w2d, #ce3311)";
export var colorSeverityOrange = "var(--color-severity-orange-rpbcus, #f89256)";
export var colorSeverityYellow = "var(--color-severity-yellow-4er6zq, #f2cd54)";
export var colorSeverityGrey = "var(--color-severity-grey-sp7qo8, #656871)";
export var colorBackgroundNotificationSeverityCritical = "var(--color-background-notification-severity-critical-0xl8pp, #870303)";
export var colorBackgroundNotificationSeverityHigh = "var(--color-background-notification-severity-high-8nbgdi, #ce3311)";
export var colorBackgroundNotificationSeverityMedium = "var(--color-background-notification-severity-medium-lbljs2, #f89256)";
export var colorBackgroundNotificationSeverityLow = "var(--color-background-notification-severity-low-giz8b6, #f2cd54)";
export var colorBackgroundNotificationSeverityNeutral = "var(--color-background-notification-severity-neutral-hnhgmv, #656871)";
export var colorTextNotificationSeverityCritical = "var(--color-text-notification-severity-critical-tv4vw4, #f9f9fa)";
export var colorTextNotificationSeverityHigh = "var(--color-text-notification-severity-high-t4suvu, #f9f9fa)";
export var colorTextNotificationSeverityMedium = "var(--color-text-notification-severity-medium-8f60kb, #0f141a)";
export var colorTextNotificationSeverityLow = "var(--color-text-notification-severity-low-gvojhi, #0f141a)";
export var colorTextNotificationSeverityNeutral = "var(--color-text-notification-severity-neutral-ynm2wl, #f9f9fa)";
export var colorGreyOpaque10 = "var(--color-grey-opaque-10-vwfmts, rgba(0, 0, 0, 0.1))";
export var colorGreyOpaque25 = "var(--color-grey-opaque-25-cjy3al, rgba(255, 255, 255, 0.25))";
export var colorGreyOpaque40 = "var(--color-grey-opaque-40-yj2cvz, rgba(0, 0, 0, 0.4))";
export var colorGreyOpaque50 = "var(--color-grey-opaque-50-dlm5ly, rgba(0, 0, 0, 0.5))";
export var colorGreyOpaque70 = "var(--color-grey-opaque-70-p0svy7, rgba(35, 43, 55, 0.7))";
export var colorGreyOpaque80 = "var(--color-grey-opaque-80-1wrh9m, rgba(22, 25, 31, 0.8))";
export var colorGreyOpaque90 = "var(--color-grey-opaque-90-te8til, rgba(242, 243, 243, 0.9))";
export var colorGreyTransparent = "var(--color-grey-transparent-g4kcvh, rgba(15, 20, 26, 0.12))";
export var colorGreyTransparentHeavy = "var(--color-grey-transparent-heavy-bu3q4i, rgba(15, 20, 26, 0.12))";
export var colorGreyTransparentLight = "var(--color-grey-transparent-light-mwvx49, rgba(15, 20, 26, 0.12))";
export var colorBackgroundBadgeIcon = "var(--color-background-badge-icon-jyxnxa, #db0000)";
export var colorBackgroundButtonLinkActive = "var(--color-background-button-link-active-5oi2dp, #d1f1ff)";
export var colorBackgroundButtonLinkHover = "var(--color-background-button-link-hover-lhrs2u, #f0fbff)";
export var colorBackgroundButtonNormalActive = "var(--color-background-button-normal-active-5imwxd, #d1f1ff)";
export var colorBackgroundButtonNormalDefault = "var(--color-background-button-normal-default-7f99mv, #ffffff)";
export var colorBackgroundButtonNormalDisabled = "var(--color-background-button-normal-disabled-hl039l, #ffffff)";
export var colorBackgroundButtonNormalHover = "var(--color-background-button-normal-hover-53op9s, #f0fbff)";
export var colorBackgroundToggleButtonNormalPressed = "var(--color-background-toggle-button-normal-pressed-4khex7, #d1f1ff)";
export var colorBackgroundButtonPrimaryActive = "var(--color-background-button-primary-active-5cqoqt, #002b66)";
export var colorBackgroundButtonPrimaryDefault = "var(--color-background-button-primary-default-vdt0fu, #006ce0)";
export var colorBackgroundButtonPrimaryDisabled = "var(--color-background-button-primary-disabled-sgo4zo, #ebebf0)";
export var colorBackgroundButtonPrimaryHover = "var(--color-background-button-primary-hover-mo85i6, #002b66)";
export var colorBackgroundDirectionButtonActive = "var(--color-background-direction-button-active-lvo0dy, #232b37)";
export var colorBackgroundDirectionButtonDefault = "var(--color-background-direction-button-default-bvhbsn, #424650)";
export var colorBackgroundDirectionButtonDisabled = "var(--color-background-direction-button-disabled-s9x4zq, #ebebf0)";
export var colorBackgroundDirectionButtonHover = "var(--color-background-direction-button-hover-74n5o1, #333843)";
export var colorTextDirectionButtonDefault = "var(--color-text-direction-button-default-p88lvb, #ffffff)";
export var colorTextDirectionButtonDisabled = "var(--color-text-direction-button-disabled-2jds36, #b4b4bb)";
export var colorBackgroundCalendarCurrentDate = "var(--color-background-calendar-current-date-sk0f6i, #f3f3f7)";
export var colorBackgroundCellShaded = "var(--color-background-cell-shaded-v7o6so, #f6f6f9)";
export var colorBackgroundCodeEditorGutterActiveLineDefault = "var(--color-background-code-editor-gutter-active-line-default-51v1pv, #656871)";
export var colorBackgroundCodeEditorGutterActiveLineError = "var(--color-background-code-editor-gutter-active-line-error-ro2qo1, #db0000)";
export var colorBackgroundCodeEditorGutterDefault = "var(--color-background-code-editor-gutter-default-15qdwh, #f3f3f7)";
export var colorBackgroundCodeEditorLoading = "var(--color-background-code-editor-loading-6nwpin, #f9f9fa)";
export var colorBackgroundCodeEditorPaneItemHover = "var(--color-background-code-editor-pane-item-hover-z6k9mr, #ebebf0)";
export var colorBackgroundCodeEditorStatusBar = "var(--color-background-code-editor-status-bar-yjtxod, #f3f3f7)";
export var colorBackgroundContainerContent = "var(--color-background-container-content-6u8rvp, #ffffff)";
export var colorBackgroundContainerHeader = "var(--color-background-container-header-gs3mbe, #ffffff)";
export var colorBackgroundControlChecked = "var(--color-background-control-checked-ka7kc2, #006ce0)";
export var colorBackgroundControlDefault = "var(--color-background-control-default-4jb21l, #ffffff)";
export var colorBackgroundControlDisabled = "var(--color-background-control-disabled-1f3718, #dedee3)";
export var colorBackgroundDropdownItemDefault = "var(--color-background-dropdown-item-default-qmc033, #ffffff)";
export var colorBackgroundDropdownItemDimmed = "var(--color-background-dropdown-item-dimmed-dhho03, transparent)";
export var colorBackgroundDropdownItemFilterMatch = "var(--color-background-dropdown-item-filter-match-49b5vt, #f0fbff)";
export var colorBackgroundDropdownItemHover = "var(--color-background-dropdown-item-hover-yunepc, #f3f3f7)";
export var colorBackgroundDropdownItemSelected = "var(--color-background-dropdown-item-selected-f3v6te, #f0fbff)";
export var colorBackgroundHomeHeader = "var(--color-background-home-header-4c9jt4, #0f141a)";
export var colorBackgroundInlineCode = "var(--color-background-inline-code-un8udy, rgba(0, 0, 0, 0.1))";
export var colorBackgroundInputDefault = "var(--color-background-input-default-ifz5bb, #ffffff)";
export var colorBackgroundInputDisabled = "var(--color-background-input-disabled-dihaja, #ebebf0)";
export var colorBackgroundItemSelected = "var(--color-background-item-selected-9gppru, #f0fbff)";
export var colorBackgroundLayoutMain = "var(--color-background-layout-main-5ilwcb, #ffffff)";
export var colorBackgroundLayoutMobilePanel = "var(--color-background-layout-mobile-panel-ed0ava, #0f141a)";
export var colorBackgroundLayoutPanelContent = "var(--color-background-layout-panel-content-xto15e, #ffffff)";
export var colorBackgroundLayoutPanelHover = "var(--color-background-layout-panel-hover-tguulw, #ebebf0)";
export var colorBackgroundLayoutToggleActive = "var(--color-background-layout-toggle-active-ap91vm, #424650)";
export var colorBackgroundLayoutToggleDefault = "var(--color-background-layout-toggle-default-2hgjdu, #424650)";
export var colorBackgroundLayoutToggleHover = "var(--color-background-layout-toggle-hover-0cpm7g, #656871)";
export var colorBackgroundLayoutToggleSelectedActive = "var(--color-background-layout-toggle-selected-active-zcl8w3, #006ce0)";
export var colorBackgroundLayoutToggleSelectedDefault = "var(--color-background-layout-toggle-selected-default-izfana, #006ce0)";
export var colorBackgroundLayoutToggleSelectedHover = "var(--color-background-layout-toggle-selected-hover-7953u1, #004a9e)";
export var colorBackgroundModalOverlay = "var(--color-background-modal-overlay-d7uby0, rgba(35, 43, 55, 0.7))";
export var colorBackgroundNotificationBlue = "var(--color-background-notification-blue-4vnob8, #006ce0)";
export var colorBackgroundNotificationGreen = "var(--color-background-notification-green-2rkyvu, #00802f)";
export var colorBackgroundNotificationGrey = "var(--color-background-notification-grey-x3vul6, #424650)";
export var colorBackgroundNotificationRed = "var(--color-background-notification-red-0487ea, #db0000)";
export var colorBackgroundNotificationYellow = "var(--color-background-notification-yellow-y6us5r, #ffe347)";
export var colorBackgroundNotificationStackBar = "var(--color-background-notification-stack-bar-qe5n4w, #232b37)";
export var colorBackgroundNotificationStackBarActive = "var(--color-background-notification-stack-bar-active-a4h9r8, #232b37)";
export var colorBackgroundNotificationStackBarHover = "var(--color-background-notification-stack-bar-hover-jh82oo, #424650)";
export var colorBackgroundPopover = "var(--color-background-popover-e20fy8, #ffffff)";
export var colorBackgroundProgressBarValueDefault = "var(--color-background-progress-bar-value-default-69ydqg, #006ce0)";
export var colorBackgroundProgressBarDefault = "var(--color-background-progress-bar-default-j8kyxd, #ebebf0)";
export var colorBackgroundSegmentActive = "var(--color-background-segment-active-1u2ldl, #006ce0)";
export var colorBackgroundSegmentDefault = "var(--color-background-segment-default-b0r494, #ffffff)";
export var colorBackgroundSegmentDisabled = "var(--color-background-segment-disabled-m2a5t7, #ffffff)";
export var colorBackgroundSegmentHover = "var(--color-background-segment-hover-800sl4, #f0fbff)";
export var colorBackgroundSegmentWrapper = "var(--color-background-segment-wrapper-5tudmm, #ffffff)";
export var colorBackgroundSliderRangeDefault = "var(--color-background-slider-range-default-3ljdu4, #006ce0)";
export var colorBackgroundSliderRangeActive = "var(--color-background-slider-range-active-vu3lky, #004a9e)";
export var colorBackgroundSliderHandleDefault = "var(--color-background-slider-handle-default-lp5ntg, #006ce0)";
export var colorBackgroundSliderHandleActive = "var(--color-background-slider-handle-active-50ubqb, #004a9e)";
export var colorBackgroundSliderTrackDefault = "var(--color-background-slider-track-default-vk2c9o, #8c8c94)";
export var colorBackgroundSliderHandleRing = "var(--color-background-slider-handle-ring-9sfenj, #ffffff)";
export var colorBackgroundSliderHandleErrorDefault = "var(--color-background-slider-handle-error-default-411tqq, #db0000)";
export var colorBackgroundSliderHandleErrorActive = "var(--color-background-slider-handle-error-active-x65pfh, #db0000)";
export var colorBackgroundSliderHandleWarningDefault = "var(--color-background-slider-handle-warning-default-or76ej, #855900)";
export var colorBackgroundSliderHandleWarningActive = "var(--color-background-slider-handle-warning-active-7o84zp, #855900)";
export var colorBackgroundSliderRangeErrorDefault = "var(--color-background-slider-range-error-default-b519qn, #db0000)";
export var colorBackgroundSliderRangeErrorActive = "var(--color-background-slider-range-error-active-6bo8an, #db0000)";
export var colorBackgroundSliderRangeWarningDefault = "var(--color-background-slider-range-warning-default-6isqvo, #855900)";
export var colorBackgroundSliderRangeWarningActive = "var(--color-background-slider-range-warning-active-dm2pha, #855900)";
export var colorBackgroundStatusError = "var(--color-background-status-error-mu3lcw, #fff5f5)";
export var colorBackgroundStatusInfo = "var(--color-background-status-info-sfobba, #f0fbff)";
export var colorBackgroundDialog = "var(--color-background-dialog-2fj3uu, #f0fbff)";
export var colorBackgroundStatusSuccess = "var(--color-background-status-success-h6b8bh, #effff1)";
export var colorBackgroundStatusWarning = "var(--color-background-status-warning-cv83up, #fffef0)";
export var colorBackgroundTableHeader = "var(--color-background-table-header-hdjxos, #ffffff)";
export var colorBackgroundTilesDisabled = "var(--color-background-tiles-disabled-4ynms7, #ebebf0)";
export var colorBackgroundToggleCheckedDisabled = "var(--color-background-toggle-checked-disabled-amxc0a, #b8e7ff)";
export var colorBackgroundToggleDefault = "var(--color-background-toggle-default-kjlhv0, #424650)";
export var colorBackgroundAvatarGenAi = "var(--color-background-avatar-gen-ai-oxp2v6, radial-gradient(circle farthest-corner at top right, #b8e7ff 0%, #0099ff 25%, #5c7fff 40% , #8575ff 60%, #962eff 80%))";
export var colorBackgroundAvatarDefault = "var(--color-background-avatar-default-t427xm, #424650)";
export var colorTextAvatar = "var(--color-text-avatar-kuhkoa, #ffffff)";
export var colorBackgroundLoadingBarGenAi = "var(--color-background-loading-bar-gen-ai-tey70i, linear-gradient(90deg, #b8e7ff 0%, #0099ff 10%, #5c7fff 24%, #8575ff 50%, #962eff 76%, #0099ff 90%, #b8e7ff 100%))";
export var colorBackgroundChatBubbleOutgoing = "var(--color-background-chat-bubble-outgoing-ay6nj3, transparent)";
export var colorBackgroundChatBubbleIncoming = "var(--color-background-chat-bubble-incoming-j38cew, #f6f6f9)";
export var colorTextChatBubbleOutgoing = "var(--color-text-chat-bubble-outgoing-f3r63s, #0f141a)";
export var colorTextChatBubbleIncoming = "var(--color-text-chat-bubble-incoming-od0yh8, #0f141a)";
export var colorBorderButtonNormalActive = "var(--color-border-button-normal-active-ru7yhb, #002b66)";
export var colorBorderButtonNormalDefault = "var(--color-border-button-normal-default-glqfp1, #006ce0)";
export var colorBorderToggleButtonNormalPressed = "var(--color-border-toggle-button-normal-pressed-tq8o41, #006ce0)";
export var colorBorderButtonNormalDisabled = "var(--color-border-button-normal-disabled-pkhetz, #b4b4bb)";
export var colorTextButtonNormalDisabled = "var(--color-text-button-normal-disabled-05p74s, #8c8c94)";
export var colorBorderButtonNormalHover = "var(--color-border-button-normal-hover-6a2tdq, #002b66)";
export var colorTextButtonIconDisabled = "var(--color-text-button-icon-disabled-nnofkn, #8c8c94)";
export var colorBorderButtonPrimaryDisabled = "var(--color-border-button-primary-disabled-b5p1ji, #ebebf0)";
export var colorTextButtonPrimaryDisabled = "var(--color-text-button-primary-disabled-q79gms, #8c8c94)";
export var colorItemSelected = "var(--color-item-selected-72rnwy, #006ce0)";
export var colorBorderCalendarGrid = "var(--color-border-calendar-grid-67r4w4, transparent)";
export var colorBorderCalendarGridSelectedFocusRing = "var(--color-border-calendar-grid-selected-focus-ring-jk1fb0, #f9f9fa)";
export var colorBorderCellShaded = "var(--color-border-cell-shaded-0ipazf, #dedee3)";
export var colorBorderCodeEditorAceActiveLineLightTheme = "var(--color-border-code-editor-ace-active-line-light-theme-q6hsvt, #dedee3)";
export var colorBorderCodeEditorAceActiveLineDarkTheme = "var(--color-border-code-editor-ace-active-line-dark-theme-v09eti, #656871)";
export var colorBorderCodeEditorDefault = "var(--color-border-code-editor-default-2bfcfq, #dedee3)";
export var colorBorderCodeEditorPaneItemHover = "var(--color-border-code-editor-pane-item-hover-wvblek, #8c8c94)";
export var colorBorderContainerDivider = "var(--color-border-container-divider-9huz1a, transparent)";
export var colorBorderContainerTop = "var(--color-border-container-top-k3vmoz, transparent)";
export var colorBorderControlChecked = "var(--color-border-control-checked-bdv28l, #006ce0)";
export var colorBorderControlDefault = "var(--color-border-control-default-sh3548, #8c8c94)";
export var colorBorderControlDisabled = "var(--color-border-control-disabled-uj7t08, #dedee3)";
export var colorBorderDividerActive = "var(--color-border-divider-active-biq3j4, #0f141a)";
export var colorBorderDividerDefault = "var(--color-border-divider-default-nr68jt, #c6c6cd)";
export var colorBorderDividerPanelBottom = "var(--color-border-divider-panel-bottom-bruvuz, #c6c6cd)";
export var colorBorderDividerPanelSide = "var(--color-border-divider-panel-side-an0w07, #c6c6cd)";
export var colorBorderDividerSecondary = "var(--color-border-divider-secondary-qoitch, #ebebf0)";
export var colorBorderDropdownContainer = "var(--color-border-dropdown-container-cmthq7, #b4b4bb)";
export var colorBorderDropdownGroup = "var(--color-border-dropdown-group-ylcnh8, #c6c6cd)";
export var colorBorderDropdownItemDefault = "var(--color-border-dropdown-item-default-kape37, #c6c6cd)";
export var colorBorderDropdownItemHover = "var(--color-border-dropdown-item-hover-aqfuxq, #8c8c94)";
export var colorBorderDropdownItemDimmedHover = "var(--color-border-dropdown-item-dimmed-hover-ga9sch, #8c8c94)";
export var colorBorderDropdownItemSelected = "var(--color-border-dropdown-item-selected-dl2ezh, #006ce0)";
export var colorBorderDropdownItemTop = "var(--color-border-dropdown-item-top-gp2d1p, transparent)";
export var colorBorderEditableCellHover = "var(--color-border-editable-cell-hover-2hmo55, #8c8c94)";
export var colorBorderInputDefault = "var(--color-border-input-default-317xk5, #8c8c94)";
export var colorBorderInputDisabled = "var(--color-border-input-disabled-zgnzvk, #ebebf0)";
export var colorBorderInputFocused = "var(--color-border-input-focused-4z0pgn, #006ce0)";
export var colorBorderItemFocused = "var(--color-border-item-focused-uk47pl, #006ce0)";
export var colorBorderDropdownItemFocused = "var(--color-border-dropdown-item-focused-zacqlp, #424650)";
export var colorBorderItemPlaceholder = "var(--color-border-item-placeholder-x8kbjp, #006ce0)";
export var colorBorderItemSelected = "var(--color-border-item-selected-wl5ttm, #006ce0)";
export var colorBorderLayout = "var(--color-border-layout-ayg8vb, #c6c6cd)";
export var colorBorderNotificationStackBar = "var(--color-border-notification-stack-bar-aszsse, #232b37)";
export var colorBorderPanelHeader = "var(--color-border-panel-header-ygztvl, #c6c6cd)";
export var colorBorderPopover = "var(--color-border-popover-1ye6tz, #b4b4bb)";
export var colorBorderSegmentActive = "var(--color-border-segment-active-ls9t4n, #424650)";
export var colorBorderSegmentDefault = "var(--color-border-segment-default-6ig2mo, #424650)";
export var colorBorderSegmentDisabled = "var(--color-border-segment-disabled-fcrbcl, #424650)";
export var colorBorderSegmentHover = "var(--color-border-segment-hover-0o9ey3, #424650)";
export var colorBorderStatusError = "var(--color-border-status-error-j8acpp, #db0000)";
export var colorBorderStatusInfo = "var(--color-border-status-info-qf6jok, #006ce0)";
export var colorBorderStatusSuccess = "var(--color-border-status-success-8z5f8u, #00802f)";
export var colorBorderStatusWarning = "var(--color-border-status-warning-j40pg7, #855900)";
export var colorBorderDialog = "var(--color-border-dialog-0rjwug, #006ce0)";
export var colorBorderDividerInteractiveDefault = "var(--color-border-divider-interactive-default-r928dz, #8c8c94)";
export var colorBorderTabsDivider = "var(--color-border-tabs-divider-f5t9va, #c6c6cd)";
export var colorBorderTabsShadow = "var(--color-border-tabs-shadow-ugyo07, rgba(15, 20, 26, 0.12))";
export var colorBorderTabsUnderline = "var(--color-border-tabs-underline-gudemr, #006ce0)";
export var colorBorderTilesDisabled = "var(--color-border-tiles-disabled-19olbu, #ebebf0)";
export var colorBorderTutorial = "var(--color-border-tutorial-zggi80, #dedee3)";
export var colorForegroundControlDefault = "var(--color-foreground-control-default-eto4wy, #ffffff)";
export var colorForegroundControlDisabled = "var(--color-foreground-control-disabled-txi6cf, #ffffff)";
export var colorForegroundControlReadOnly = "var(--color-foreground-control-read-only-7ydvuj, #656871)";
export var colorShadowDefault = "var(--color-shadow-default-o7dmmm, rgba(15, 20, 26, 0.12))";
export var colorShadowMedium = "var(--color-shadow-medium-x7of55, rgba(15, 20, 26, 0.12))";
export var colorShadowSide = "var(--color-shadow-side-rtsbda, rgba(15, 20, 26, 0.12))";
export var colorStrokeChartLine = "var(--color-stroke-chart-line-3nsnk6, #8c8c94)";
export var colorStrokeCodeEditorGutterActiveLineDefault = "var(--color-stroke-code-editor-gutter-active-line-default-5hrdmu, #dedee3)";
export var colorStrokeCodeEditorGutterActiveLineHover = "var(--color-stroke-code-editor-gutter-active-line-hover-vqrb6s, #f9f9fa)";
export var colorTextAccent = "var(--color-text-accent-n1kmht, #006ce0)";
export var colorTextBodyDefault = "var(--color-text-body-default-vvtq8u, #0f141a)";
export var colorTextBodySecondary = "var(--color-text-body-secondary-yna5sb, #424650)";
export var colorTextBreadcrumbCurrent = "var(--color-text-breadcrumb-current-2mqnkk, #656871)";
export var colorTextBreadcrumbIcon = "var(--color-text-breadcrumb-icon-9j48ot, #8c8c94)";
export var colorTextButtonInlineIconDefault = "var(--color-text-button-inline-icon-default-sm4ql6, #006ce0)";
export var colorTextButtonInlineIconDisabled = "var(--color-text-button-inline-icon-disabled-82hho0, #b4b4bb)";
export var colorTextButtonInlineIconHover = "var(--color-text-button-inline-icon-hover-rbyzfc, #002b66)";
export var colorTextButtonNormalActive = "var(--color-text-button-normal-active-vihsxh, #002b66)";
export var colorTextToggleButtonNormalPressed = "var(--color-text-toggle-button-normal-pressed-wnx2zl, #002b66)";
export var colorTextButtonNormalDefault = "var(--color-text-button-normal-default-nzalii, #006ce0)";
export var colorTextButtonNormalHover = "var(--color-text-button-normal-hover-gusgyv, #002b66)";
export var colorTextLinkButtonNormalDefault = "var(--color-text-link-button-normal-default-srprth, #006ce0)";
export var colorTextLinkButtonNormalHover = "var(--color-text-link-button-normal-hover-jrnyw3, #002b66)";
export var colorTextLinkButtonNormalActive = "var(--color-text-link-button-normal-active-js9ryu, #002b66)";
export var colorTextButtonPrimaryActive = "var(--color-text-button-primary-active-refmba, #ffffff)";
export var colorTextButtonPrimaryDefault = "var(--color-text-button-primary-default-mwl31m, #ffffff)";
export var colorTextButtonPrimaryHover = "var(--color-text-button-primary-hover-pw12ep, #ffffff)";
export var colorTextCalendarDateHover = "var(--color-text-calendar-date-hover-3fcriv, #0f141a)";
export var colorTextCalendarMonth = "var(--color-text-calendar-month-ea0e93, #656871)";
export var colorTextCodeEditorGutterActiveLine = "var(--color-text-code-editor-gutter-active-line-2addhd, #ffffff)";
export var colorTextCodeEditorGutterDefault = "var(--color-text-code-editor-gutter-default-nlshs8, #0f141a)";
export var colorTextCodeEditorStatusBarDisabled = "var(--color-text-code-editor-status-bar-disabled-xxmtlc, #8c8c94)";
export var colorTextCodeEditorTabButtonError = "var(--color-text-code-editor-tab-button-error-avwh01, #ffffff)";
export var colorTextColumnHeader = "var(--color-text-column-header-e6urd1, #424650)";
export var colorTextColumnSortingIcon = "var(--color-text-column-sorting-icon-fngn77, #424650)";
export var colorTextControlDisabled = "var(--color-text-control-disabled-upk9lz, #b4b4bb)";
export var colorTextCounter = "var(--color-text-counter-bywf75, #656871)";
export var colorTextDisabled = "var(--color-text-disabled-rox5hg, #b4b4bb)";
export var colorTextDisabledInlineEdit = "var(--color-text-disabled-inline-edit-1cbiz8, #424650)";
export var colorTextDropdownFooter = "var(--color-text-dropdown-footer-umcot2, #656871)";
export var colorTextDropdownGroupLabel = "var(--color-text-dropdown-group-label-2tmyik, #424650)";
export var colorTextDropdownItemDefault = "var(--color-text-dropdown-item-default-f1jr9u, #0f141a)";
export var colorTextDropdownItemDimmed = "var(--color-text-dropdown-item-dimmed-tq8vh3, #b4b4bb)";
export var colorTextDropdownItemDisabled = "var(--color-text-dropdown-item-disabled-8m65hf, #b4b4bb)";
export var colorTextDropdownItemFilterMatch = "var(--color-text-dropdown-item-filter-match-ebhvct, #006ce0)";
export var colorTextDropdownItemHighlighted = "var(--color-text-dropdown-item-highlighted-yr1px8, #0f141a)";
export var colorTextDropdownItemSecondary = "var(--color-text-dropdown-item-secondary-v12lfh, #656871)";
export var colorTextDropdownItemSecondaryHover = "var(--color-text-dropdown-item-secondary-hover-de15wb, #656871)";
export var colorTextEmpty = "var(--color-text-empty-tlohug, #656871)";
export var colorTextExpandableSectionDefault = "var(--color-text-expandable-section-default-ynw8my, #0f141a)";
export var colorTextExpandableSectionHover = "var(--color-text-expandable-section-hover-ojzwhd, #006ce0)";
export var colorTextExpandableSectionNavigationIconDefault = "var(--color-text-expandable-section-navigation-icon-default-mklu1s, #424650)";
export var colorTextFormDefault = "var(--color-text-form-default-47mtz6, #0f141a)";
export var colorTextFormLabel = "var(--color-text-form-label-6sbm75, #0f141a)";
export var colorTextFormSecondary = "var(--color-text-form-secondary-1nm780, #656871)";
export var colorTextGroupLabel = "var(--color-text-group-label-a2qc05, #424650)";
export var colorTextLabelGenAi = "var(--color-text-label-gen-ai-a2n3od, #7300e5)";
export var colorTextHeadingDefault = "var(--color-text-heading-default-izpp46, #0f141a)";
export var colorTextHeadingSecondary = "var(--color-text-heading-secondary-iwtvf6, #424650)";
export var colorTextHomeHeaderDefault = "var(--color-text-home-header-default-morg6i, #ebebf0)";
export var colorTextHomeHeaderSecondary = "var(--color-text-home-header-secondary-i4jhp7, #c6c6cd)";
export var colorTextIconCaret = "var(--color-text-icon-caret-9mqubk, #8c8c94)";
export var colorTextIconSubtle = "var(--color-text-icon-subtle-3sgxlr, #656871)";
export var colorTextInputDisabled = "var(--color-text-input-disabled-wh1f3y, #b4b4bb)";
export var colorTextInputPlaceholder = "var(--color-text-input-placeholder-dclg8u, #656871)";
export var colorTextInputPlaceholderDisabled = "var(--color-text-input-placeholder-disabled-wg87og, #b4b4bb)";
export var colorTextInteractiveActive = "var(--color-text-interactive-active-uoe6zi, #0f141a)";
export var colorTextInteractiveDefault = "var(--color-text-interactive-default-ugh9wp, #424650)";
export var colorTextInteractiveDisabled = "var(--color-text-interactive-disabled-1bqmrl, #b4b4bb)";
export var colorTextInteractiveHover = "var(--color-text-interactive-hover-6naf7i, #0f141a)";
export var colorTextToggleButtonIconPressed = "var(--color-text-toggle-button-icon-pressed-detfkz, #0f141a)";
export var colorTextInteractiveInvertedDefault = "var(--color-text-interactive-inverted-default-xlc0d5, #dedee3)";
export var colorTextInteractiveInvertedHover = "var(--color-text-interactive-inverted-hover-65rnp7, #f9f9fa)";
export var colorTextInverted = "var(--color-text-inverted-4v4dmq, #ffffff)";
export var colorTextLabel = "var(--color-text-label-28gfmc, #0f141a)";
export var colorTextLayoutToggle = "var(--color-text-layout-toggle-1a15s3, #ffffff)";
export var colorTextLayoutToggleActive = "var(--color-text-layout-toggle-active-ifu7qp, #ffffff)";
export var colorTextLayoutToggleHover = "var(--color-text-layout-toggle-hover-9jwdce, #006ce0)";
export var colorTextLayoutToggleSelected = "var(--color-text-layout-toggle-selected-xpximc, #ffffff)";
export var colorTextLinkDefault = "var(--color-text-link-default-hude44, #006ce0)";
export var colorTextLinkHover = "var(--color-text-link-hover-2hfec2, #002b66)";
export var colorTextLinkInvertedHover = "var(--color-text-link-inverted-hover-ocd3u3, #ffffff)";
export var colorTextLinkButtonUnderline = "var(--color-text-link-button-underline-z4wjnv, transparent)";
export var colorTextLinkButtonUnderlineHover = "var(--color-text-link-button-underline-hover-cn3mqh, transparent)";
export var colorTextNotificationDefault = "var(--color-text-notification-default-1iey72, #f9f9fa)";
export var colorTextNotificationStackBar = "var(--color-text-notification-stack-bar-tjj0ek, #ffffff)";
export var colorTextNotificationYellow = "var(--color-text-notification-yellow-vjtdxk, #0f141a)";
export var colorTextPaginationPageNumberActiveDisabled = "var(--color-text-pagination-page-number-active-disabled-gfl43p, #b4b4bb)";
export var colorTextPaginationPageNumberDefault = "var(--color-text-pagination-page-number-default-74j15c, #424650)";
export var colorTextSegmentActive = "var(--color-text-segment-active-hlorbe, #ffffff)";
export var colorTextSegmentDefault = "var(--color-text-segment-default-vi2vn9, #424650)";
export var colorTextSegmentHover = "var(--color-text-segment-hover-65a2x8, #002b66)";
export var colorTextSmall = "var(--color-text-small-m1tr70, #656871)";
export var colorTextStatusError = "var(--color-text-status-error-ksqavh, #db0000)";
export var colorTextStatusInactive = "var(--color-text-status-inactive-gy7337, #656871)";
export var colorTextStatusInfo = "var(--color-text-status-info-ue8bd2, #006ce0)";
export var colorTextStatusSuccess = "var(--color-text-status-success-ybmii8, #00802f)";
export var colorTextStatusWarning = "var(--color-text-status-warning-6meo06, #855900)";
export var colorTextTopNavigationTitle = "var(--color-text-top-navigation-title-en0v40, #0f141a)";
export var colorTextTutorialHotspotDefault = "var(--color-text-tutorial-hotspot-default-xfv3ow, #006ce0)";
export var colorTextTutorialHotspotHover = "var(--color-text-tutorial-hotspot-hover-92pxog, #002b66)";
export var colorBoardPlaceholderActive = "var(--color-board-placeholder-active-x6yfem, #ebebf0)";
export var colorBoardPlaceholderHover = "var(--color-board-placeholder-hover-5nov9c, #d1f1ff)";
export var colorDragPlaceholderActive = "var(--color-drag-placeholder-active-ea1sgp, #ebebf0)";
export var colorDragPlaceholderHover = "var(--color-drag-placeholder-hover-qwadna, #d1f1ff)";
export var colorDropzoneBackgroundDefault = "var(--color-dropzone-background-default-7efhmb, #ffffff)";
export var colorDropzoneBackgroundHover = "var(--color-dropzone-background-hover-mi8rlm, #f0fbff)";
export var colorDropzoneTextDefault = "var(--color-dropzone-text-default-djdnme, #424650)";
export var colorDropzoneTextHover = "var(--color-dropzone-text-hover-asw3rt, #424650)";
export var colorDropzoneBorderDefault = "var(--color-dropzone-border-default-k648ha, #8c8c94)";
export var colorDropzoneBorderHover = "var(--color-dropzone-border-hover-otpag5, #002b66)";
export var colorGapGlobalDrawer = "var(--color-gap-global-drawer-nh699a, #ebebf0)";
export var colorTreeViewConnectorLine = "var(--color-tree-view-connector-line-1usxvn, #8c8c94)";
export var fontBoxValueLargeWeight = "var(--font-box-value-large-weight-wr00sw, 700)";
export var fontButtonLetterSpacing = "var(--font-button-letter-spacing-ufowe3, 0.005em)";
export var fontChartDetailSize = "var(--font-chart-detail-size-9qr25q, 12px)";
export var fontDisplayLabelWeight = "var(--font-display-label-weight-zavpeo, 700)";
export var fontExpandableHeadingSize = "var(--font-expandable-heading-size-0uk059, 16px)";
export var fontFamilyBase = "var(--font-family-base-gmnpzl, 'Open Sans', 'Helvetica Neue', Roboto, Arial, sans-serif)";
export var fontFamilyMonospace = "var(--font-family-monospace-q47m7k, Monaco, Menlo, Consolas, 'Courier Prime', Courier, 'Courier New', monospace)";
export var fontHeaderH2DescriptionLineHeight = "var(--font-header-h2-description-line-height-ts2s6o, 20px)";
export var fontHeaderH2DescriptionSize = "var(--font-header-h2-description-size-g2wws3, 14px)";
export var fontLinkButtonLetterSpacing = "var(--font-link-button-letter-spacing-imtxwq, 0.005em)";
export var fontLinkButtonWeight = "var(--font-link-button-weight-vslyg9, 700)";
export var fontPanelHeaderLineHeight = "var(--font-panel-header-line-height-8xb2qj, 22px)";
export var fontPanelHeaderSize = "var(--font-panel-header-size-33h9j8, 18px)";
export var fontSizeBodyM = "var(--font-size-body-m-a7nh2n, 14px)";
export var fontSizeBodyS = "var(--font-size-body-s-smc8cv, 12px)";
export var fontSizeDisplayL = "var(--font-size-display-l-wa6woo, 42px)";
export var fontSizeHeadingXl = "var(--font-size-heading-xl-wvkbur, 24px)";
export var fontSizeHeadingL = "var(--font-size-heading-l-vnacx6, 20px)";
export var fontSizeHeadingM = "var(--font-size-heading-m-170yiy, 18px)";
export var fontSizeHeadingS = "var(--font-size-heading-s-zp08en, 16px)";
export var fontSizeHeadingXs = "var(--font-size-heading-xs-j8yzxv, 14px)";
export var fontSmoothingMozOsx = "var(--font-smoothing-moz-osx-hbm0aq, grayscale)";
export var fontSmoothingWebkit = "var(--font-smoothing-webkit-oemolo, antialiased)";
export var fontTabsDisabledWeight = "var(--font-tabs-disabled-weight-e3lt4d, 700)";
export var fontTabsLineHeight = "var(--font-tabs-line-height-5g6n1v, 20px)";
export var fontTabsSize = "var(--font-tabs-size-18irgl, 16px)";
export var fontWayfindingLinkActiveWeight = "var(--font-wayfinding-link-active-weight-ny4hup, 700)";
export var fontWeightButton = "var(--font-weight-button-0eg20c, 700)";
export var fontWeightHeadingXl = "var(--font-weight-heading-xl-u3m4we, 700)";
export var fontWeightHeadingL = "var(--font-weight-heading-l-0t6dwc, 700)";
export var fontWeightHeadingM = "var(--font-weight-heading-m-zf82dr, 700)";
export var fontWeightHeadingS = "var(--font-weight-heading-s-lcx0ai, 700)";
export var fontWeightHeadingXs = "var(--font-weight-heading-xs-wqqpne, 700)";
export var fontWeightHeavy = "var(--font-weight-heavy-6yh4un, 700)";
export var letterSpacingBodyS = "var(--letter-spacing-body-s-gq78ok, 0.005em)";
export var letterSpacingDisplayL = "var(--letter-spacing-display-l-elyyxk, -0.03em)";
export var letterSpacingHeadingXl = "var(--letter-spacing-heading-xl-ckkb6u, -0.02em)";
export var letterSpacingHeadingL = "var(--letter-spacing-heading-l-5v6ibv, -0.015em)";
export var letterSpacingHeadingM = "var(--letter-spacing-heading-m-29ewnk, -0.010em)";
export var letterSpacingHeadingS = "var(--letter-spacing-heading-s-4st9ep, -0.005em)";
export var lineHeightBodyM = "var(--line-height-body-m-2mh3ke, 20px)";
export var lineHeightBodyS = "var(--line-height-body-s-nu5hx1, 16px)";
export var lineHeightDisplayL = "var(--line-height-display-l-vwanzp, 48px)";
export var lineHeightHeadingXl = "var(--line-height-heading-xl-hko6p0, 30px)";
export var lineHeightHeadingL = "var(--line-height-heading-l-mg5bx6, 24px)";
export var lineHeightHeadingM = "var(--line-height-heading-m-uoaqdh, 22px)";
export var lineHeightHeadingS = "var(--line-height-heading-s-hmi4vc, 20px)";
export var lineHeightHeadingXs = "var(--line-height-heading-xs-q9j004, 18px)";
export var borderActiveWidth = "var(--border-active-width-axzm24, 4px)";
export var borderCodeEditorStatusDividerWidth = "var(--border-code-editor-status-divider-width-4we6jf, 1px)";
export var borderContainerStickyWidth = "var(--border-container-sticky-width-nri0ix, 0px)";
export var borderContainerTopWidth = "var(--border-container-top-width-n1eke6, 0px)";
export var borderControlFocusRingShadowSpread = "var(--border-control-focus-ring-shadow-spread-9mjajk, 1px)";
export var borderControlInvalidFocusRingShadowSpread = "var(--border-control-invalid-focus-ring-shadow-spread-9jjf96, 2px)";
export var borderDividerListWidth = "var(--border-divider-list-width-tdfx1x, 1px)";
export var borderDividerSectionWidth = "var(--border-divider-section-width-uwo8my, 1px)";
export var borderDropdownVirtualOffsetWidth = "var(--border-dropdown-virtual-offset-width-3wp954, 2px)";
export var borderInvalidWidth = "var(--border-invalid-width-3xd6e1, 8px)";
export var borderItemWidth = "var(--border-item-width-miijiw, 2px)";
export var borderLineChartDashArray = "var(--border-line-chart-dash-array-desefi, 3 5)";
export var borderLineChartLineJoin = "var(--border-line-chart-line-join-aslwou, round)";
export var borderLineChartWidth = "var(--border-line-chart-width-tesor1, 2px)";
export var borderPanelHeaderWidth = "var(--border-panel-header-width-t1iq1m, 1px)";
export var borderPanelTopWidth = "var(--border-panel-top-width-10990j, 1px)";
export var borderRadiusAlert = "var(--border-radius-alert-syagf6, 12px)";
export var borderRadiusBadge = "var(--border-radius-badge-exolfb, 4px)";
export var borderRadiusButton = "var(--border-radius-button-7bgkcs, 20px)";
export var borderRadiusCalendarDayFocusRing = "var(--border-radius-calendar-day-focus-ring-xvvbuc, 3px)";
export var borderRadiusCodeEditor = "var(--border-radius-code-editor-5palck, 8px)";
export var borderRadiusContainer = "var(--border-radius-container-nsfwmm, 16px)";
export var borderRadiusControlCircularFocusRing = "var(--border-radius-control-circular-focus-ring-yjhscw, 4px)";
export var borderRadiusControlDefaultFocusRing = "var(--border-radius-control-default-focus-ring-1uabki, 4px)";
export var borderRadiusDropdown = "var(--border-radius-dropdown-fgc2a1, 8px)";
export var borderRadiusDropzone = "var(--border-radius-dropzone-eklq14, 12px)";
export var borderRadiusFlashbar = "var(--border-radius-flashbar-pp1ptu, 12px)";
export var borderRadiusItem = "var(--border-radius-item-iwaia5, 8px)";
export var borderRadiusInput = "var(--border-radius-input-7q0str, 8px)";
export var borderRadiusPopover = "var(--border-radius-popover-6fqb5w, 8px)";
export var borderRadiusTabsFocusRing = "var(--border-radius-tabs-focus-ring-o4qku1, 20px)";
export var borderRadiusTiles = "var(--border-radius-tiles-wm1vgw, 8px)";
export var borderRadiusToken = "var(--border-radius-token-ycnemh, 8px)";
export var borderRadiusChatBubble = "var(--border-radius-chat-bubble-haafsg, 8px)";
export var borderRadiusTutorialPanelItem = "var(--border-radius-tutorial-panel-item-ojaqxg, 8px)";
export var borderTableStickyWidth = "var(--border-table-sticky-width-ai31mi, 1px)";
export var borderLinkFocusRingOutline = "var(--border-link-focus-ring-outline-1p0hnu, 0)";
export var borderLinkFocusRingShadowSpread = "var(--border-link-focus-ring-shadow-spread-39uvxr, 2px)";
export var borderWidthAlert = "var(--border-width-alert-tuifgy, 2px)";
export var borderWidthButton = "var(--border-width-button-jm0qg7, 2px)";
export var borderWidthDropdown = "var(--border-width-dropdown-youcay, 2px)";
export var borderWidthField = "var(--border-width-field-2xc78x, 1px)";
export var borderWidthPopover = "var(--border-width-popover-nflirh, 2px)";
export var borderWidthToken = "var(--border-width-token-2ukdpu, 2px)";
export var borderWidthIconSmall = "var(--border-width-icon-small-z55i5t, 2px)";
export var borderWidthIconNormal = "var(--border-width-icon-normal-9h7vj7, 2px)";
export var borderWidthIconMedium = "var(--border-width-icon-medium-b7icqv, 2px)";
export var borderWidthIconBig = "var(--border-width-icon-big-ymgy42, 3px)";
export var borderWidthIconLarge = "var(--border-width-icon-large-u645rg, 4px)";
export var motionDurationExtraFast = "var(--motion-duration-extra-fast-l4w48j, 45ms)";
export var motionDurationExtraSlow = "var(--motion-duration-extra-slow-29bqym, 270ms)";
export var motionDurationFast = "var(--motion-duration-fast-unntf6, 90ms)";
export var motionDurationModerate = "var(--motion-duration-moderate-c9utmg, 135ms)";
export var motionDurationRefreshOnlyAmbient = "var(--motion-duration-refresh-only-ambient-sxpcba, 2000ms)";
export var motionDurationRefreshOnlyFast = "var(--motion-duration-refresh-only-fast-zfibh6, 115ms)";
export var motionDurationRefreshOnlyMedium = "var(--motion-duration-refresh-only-medium-5rbn3k, 165ms)";
export var motionDurationRefreshOnlySlow = "var(--motion-duration-refresh-only-slow-ugjy90, 250ms)";
export var motionDurationAvatarGenAiGradient = "var(--motion-duration-avatar-gen-ai-gradient-84si5n, 3600ms)";
export var motionDurationAvatarLoadingDots = "var(--motion-duration-avatar-loading-dots-1xxvis, 1200ms)";
export var motionDurationRotate180 = "var(--motion-duration-rotate-180-cxi9g7, 135ms)";
export var motionDurationRotate90 = "var(--motion-duration-rotate-90-lyzb0k, 135ms)";
export var motionDurationShowPaced = "var(--motion-duration-show-paced-otsjh8, 180ms)";
export var motionDurationShowQuick = "var(--motion-duration-show-quick-tyvnyw, 135ms)";
export var motionDurationSlow = "var(--motion-duration-slow-zji5vl, 180ms)";
export var motionDurationTransitionQuick = "var(--motion-duration-transition-quick-mcm2y0, 90ms)";
export var motionDurationTransitionShowPaced = "var(--motion-duration-transition-show-paced-t8d1os, 180ms)";
export var motionDurationTransitionShowQuick = "var(--motion-duration-transition-show-quick-5jnnjz, 90ms)";
export var motionEasingEaseOutQuart = "var(--motion-easing-ease-out-quart-p9axhm, cubic-bezier(0.165, 0.84, 0.44, 1))";
export var motionEasingRefreshOnlyA = "var(--motion-easing-refresh-only-a-ccyqaz, cubic-bezier(0, 0, 0, 1))";
export var motionEasingRefreshOnlyB = "var(--motion-easing-refresh-only-b-44kz4o, cubic-bezier(1, 0, 0.83, 1))";
export var motionEasingRefreshOnlyC = "var(--motion-easing-refresh-only-c-cxy2sk, cubic-bezier(0.84, 0, 0.16, 1))";
export var motionEasingRefreshOnlyD = "var(--motion-easing-refresh-only-d-syj3g1, cubic-bezier(0.33, 0, 0.67, 1))";
export var motionEasingAvatarGenAiGradient = "var(--motion-easing-avatar-gen-ai-gradient-9fwaak, cubic-bezier(0.7, 0, 0.3, 1))";
export var motionEasingRotate180 = "var(--motion-easing-rotate-180-7a58rc, cubic-bezier(0.165, 0.84, 0.44, 1))";
export var motionEasingRotate90 = "var(--motion-easing-rotate-90-jhbqg9, cubic-bezier(0.165, 0.84, 0.44, 1))";
export var motionEasingShowPaced = "var(--motion-easing-show-paced-ym6eyn, ease-out)";
export var motionEasingShowQuick = "var(--motion-easing-show-quick-9hlj8q, ease-out)";
export var motionEasingTransitionQuick = "var(--motion-easing-transition-quick-qxak3i, linear)";
export var motionEasingTransitionShowPaced = "var(--motion-easing-transition-show-paced-x2k7uh, ease-out)";
export var motionEasingTransitionShowQuick = "var(--motion-easing-transition-show-quick-jz3lia, linear)";
export var motionEasingResponsive = "var(--motion-easing-responsive-hjj3ai, cubic-bezier(0, 0, 0, 1))";
export var motionEasingSticky = "var(--motion-easing-sticky-tn072u, cubic-bezier(1, 0, 0.83, 1))";
export var motionEasingExpressive = "var(--motion-easing-expressive-o5jqzg, cubic-bezier(0.84, 0, 0.16, 1))";
export var motionDurationResponsive = "var(--motion-duration-responsive-mehora, 115ms)";
export var motionDurationExpressive = "var(--motion-duration-expressive-cbdcwy, 165ms)";
export var motionDurationComplex = "var(--motion-duration-complex-tbdo30, 250ms)";
export var motionKeyframesFadeIn = "var(--motion-keyframes-fade-in-0r842q, awsui-fade-in-35003c)";
export var motionKeyframesFadeOut = "var(--motion-keyframes-fade-out-g7fgdu, awsui-fade-out-35003c)";
export var motionKeyframesStatusIconError = "var(--motion-keyframes-status-icon-error-wkou39, awsui-status-icon-error-35003c)";
export var motionKeyframesScalePopup = "var(--motion-keyframes-scale-popup-9iqcu0, awsui-scale-popup-35003c)";
export var sizeCalendarGridWidth = "var(--size-calendar-grid-width-hv3136, 238px)";
export var sizeControl = "var(--size-control-adm93y, 16px)";
export var sizeIconBig = "var(--size-icon-big-7pq9l3, 32px)";
export var sizeIconLarge = "var(--size-icon-large-mb6y6y, 48px)";
export var sizeIconMedium = "var(--size-icon-medium-uv8xcz, 20px)";
export var sizeIconNormal = "var(--size-icon-normal-levt08, 16px)";
export var sizeTableSelectionHorizontal = "var(--size-table-selection-horizontal-qqiajd, 40px)";
export var sizeVerticalInput = "var(--size-vertical-input-p1d7xx, 32px)";
export var sizeVerticalPanelIconOffset = "var(--size-vertical-panel-icon-offset-z959cw, 15px)";
export var spaceAlertActionLeft = "var(--space-alert-action-left-4s8zo5, 12px)";
export var spaceAlertHorizontal = "var(--space-alert-horizontal-ul364s, 16px)";
export var spaceAlertMessageRight = "var(--space-alert-message-right-mrjbnn, 4px)";
export var spaceAlertVertical = "var(--space-alert-vertical-dlp5wr, 8px)";
export var spaceButtonFocusOutlineGutter = "var(--space-button-focus-outline-gutter-jj138g, 4px)";
export var spaceButtonHorizontal = "var(--space-button-horizontal-k0c786, 20px)";
export var spaceButtonIconFocusOutlineGutterVertical = "var(--space-button-icon-focus-outline-gutter-vertical-r44mtq, 0px)";
export var spaceButtonIconOnlyHorizontal = "var(--space-button-icon-only-horizontal-i85hxi, 6px)";
export var spaceButtonInlineIconFocusOutlineGutter = "var(--space-button-inline-icon-focus-outline-gutter-zbfgku, 0px)";
export var spaceButtonModalDismissVertical = "var(--space-button-modal-dismiss-vertical-vqfxjd, 2px)";
export var spaceCalendarGridFocusOutlineGutter = "var(--space-calendar-grid-focus-outline-gutter-vvh43m, -5px)";
export var spaceCalendarGridSelectedFocusOutlineGutter = "var(--space-calendar-grid-selected-focus-outline-gutter-dy6gf8, -5px)";
export var spaceCalendarGridGutter = "var(--space-calendar-grid-gutter-zojo6r, 6px)";
export var spaceCardHorizontal = "var(--space-card-horizontal-8o8t31, 20px)";
export var spaceCardVertical = "var(--space-card-vertical-uad0cc, 16px)";
export var spaceCodeEditorStatusFocusOutlineGutter = "var(--space-code-editor-status-focus-outline-gutter-o87hra, -7px)";
export var spaceContainerContentTop = "var(--space-container-content-top-1wtqrc, 4px)";
export var spaceContainerHeaderTop = "var(--space-container-header-top-am4vzw, 12px)";
export var spaceContainerHeaderBottom = "var(--space-container-header-bottom-2taq8v, 8px)";
export var spaceContainerHorizontal = "var(--space-container-horizontal-nqrzyh, 20px)";
export var spaceContentHeaderPaddingBottom = "var(--space-content-header-padding-bottom-rvy5xz, 16px)";
export var spaceDarkHeaderOverlapDistance = "var(--space-dark-header-overlap-distance-ld45ap, 36px)";
export var spaceExpandableSectionIconOffsetTop = "var(--space-expandable-section-icon-offset-top-cntyn8, 4px)";
export var spaceFieldHorizontal = "var(--space-field-horizontal-0aq2ch, 12px)";
export var spaceFieldIconOffset = "var(--space-field-icon-offset-ikwzwx, 36px)";
export var spaceFilteringTokenDismissButtonFocusOutlineGutter = "var(--space-filtering-token-dismiss-button-focus-outline-gutter-1iumy3, -5px)";
export var spaceFilteringTokenOperationSelectFocusOutlineGutter = "var(--space-filtering-token-operation-select-focus-outline-gutter-jacx1t, -5px)";
export var spaceFlashbarActionLeft = "var(--space-flashbar-action-left-rqk3ap, 12px)";
export var spaceFlashbarDismissRight = "var(--space-flashbar-dismiss-right-ckhj91, 0px)";
export var spaceFlashbarHorizontal = "var(--space-flashbar-horizontal-l63501, 16px)";
export var spaceFlashbarVertical = "var(--space-flashbar-vertical-th71op, 8px)";
export var spaceGridGutter = "var(--space-grid-gutter-whc3jp, 20px)";
export var spaceKeyValueGap = "var(--space-key-value-gap-9glmqc, 0px)";
export var spaceLayoutContentBottom = "var(--space-layout-content-bottom-zeb1g9, 40px)";
export var spaceLayoutContentHorizontal = "var(--space-layout-content-horizontal-buc0zz, 24px)";
export var spaceLayoutToggleDiameter = "var(--space-layout-toggle-diameter-j2qffw, 36px)";
export var spaceLayoutTogglePadding = "var(--space-layout-toggle-padding-chwlhz, 12px)";
export var spaceModalContentBottom = "var(--space-modal-content-bottom-nl6ceq, 16px)";
export var spaceModalHorizontal = "var(--space-modal-horizontal-y5hnwp, 20px)";
export var spacePanelContentBottom = "var(--space-panel-content-bottom-24c6lu, 40px)";
export var spacePanelContentTop = "var(--space-panel-content-top-qvd1dr, 20px)";
export var spacePanelDividerMarginHorizontal = "var(--space-panel-divider-margin-horizontal-yw31p0, 8px)";
export var spacePanelHeaderVertical = "var(--space-panel-header-vertical-ckfgmy, 20px)";
export var spacePanelNavLeft = "var(--space-panel-nav-left-wn0n7h, 28px)";
export var spacePanelSideLeft = "var(--space-panel-side-left-u1m3s9, 28px)";
export var spacePanelSideRight = "var(--space-panel-side-right-8wwirc, 24px)";
export var spacePanelSplitTop = "var(--space-panel-split-top-3u4vky, 20px)";
export var spacePanelSplitBottom = "var(--space-panel-split-bottom-ir16d7, 20px)";
export var spaceSegmentedControlFocusOutlineGutter = "var(--space-segmented-control-focus-outline-gutter-x1ywqb, 6px)";
export var spaceTabsContentTop = "var(--space-tabs-content-top-ju6qox, 12px)";
export var spaceTabsFocusOutlineGutter = "var(--space-tabs-focus-outline-gutter-eerrg4, -8px)";
export var spaceTableContentBottom = "var(--space-table-content-bottom-tlfqmq, 4px)";
export var spaceTableEmbeddedHeaderTop = "var(--space-table-embedded-header-top-twu628, 0px)";
export var spaceTableFooterHorizontal = "var(--space-table-footer-horizontal-l5g495, 0px)";
export var spaceTableHeaderFocusOutlineGutter = "var(--space-table-header-focus-outline-gutter-ymwujm, 0px)";
export var spaceTableHeaderHorizontal = "var(--space-table-header-horizontal-kb5ww2, 0px)";
export var spaceTableHeaderToolsBottom = "var(--space-table-header-tools-bottom-d9u5kf, 0px)";
export var spaceTableHeaderToolsFullPageBottom = "var(--space-table-header-tools-full-page-bottom-9m47g6, 4px)";
export var spaceTableHorizontal = "var(--space-table-horizontal-suurzj, 20px)";
export var spaceTreeViewIndentation = "var(--space-tree-view-indentation-xh9kis, 24px)";
export var spaceTileGutter = "var(--space-tile-gutter-bi2bdv, 24px)";
export var spaceScaled2xNone = "var(--space-scaled-2x-none-987dp7, 0px)";
export var spaceScaled2xXxxs = "var(--space-scaled-2x-xxxs-reumxj, 2px)";
export var spaceScaled2xXxs = "var(--space-scaled-2x-xxs-e79hr1, 4px)";
export var spaceScaled2xXs = "var(--space-scaled-2x-xs-bcbsqo, 8px)";
export var spaceScaled2xS = "var(--space-scaled-2x-s-yr27d5, 12px)";
export var spaceScaled2xM = "var(--space-scaled-2x-m-4euqsk, 16px)";
export var spaceScaled2xL = "var(--space-scaled-2x-l-u5ida5, 20px)";
export var spaceScaled2xXl = "var(--space-scaled-2x-xl-he48nr, 24px)";
export var spaceScaled2xXxl = "var(--space-scaled-2x-xxl-sul5ey, 32px)";
export var spaceScaled2xXxxl = "var(--space-scaled-2x-xxxl-bxyvwl, 40px)";
export var spaceScaledNone = "var(--space-scaled-none-nfyouv, 0px)";
export var spaceScaledXxxs = "var(--space-scaled-xxxs-oo06c7, 2px)";
export var spaceScaledXxs = "var(--space-scaled-xxs-pfm1nx, 4px)";
export var spaceScaledXs = "var(--space-scaled-xs-xwoogq, 8px)";
export var spaceScaledS = "var(--space-scaled-s-8ozaad, 12px)";
export var spaceScaledM = "var(--space-scaled-m-m892r9, 16px)";
export var spaceScaledL = "var(--space-scaled-l-sej05l, 20px)";
export var spaceScaledXl = "var(--space-scaled-xl-dunxp5, 24px)";
export var spaceScaledXxl = "var(--space-scaled-xxl-6wgq96, 32px)";
export var spaceScaledXxxl = "var(--space-scaled-xxxl-hwoy7j, 40px)";
export var spaceStaticXxxs = "var(--space-static-xxxs-yidks1, 2px)";
export var spaceStaticXxs = "var(--space-static-xxs-ns94dp, 4px)";
export var spaceStaticXs = "var(--space-static-xs-gnm0mz, 8px)";
export var spaceStaticS = "var(--space-static-s-t763lu, 12px)";
export var spaceStaticM = "var(--space-static-m-m6qboo, 16px)";
export var spaceStaticL = "var(--space-static-l-n53k41, 20px)";
export var spaceStaticXl = "var(--space-static-xl-4tedi6, 24px)";
export var spaceStaticXxl = "var(--space-static-xxl-ifa9j8, 32px)";
export var spaceStaticXxxl = "var(--space-static-xxxl-tngnnz, 40px)";
export var spaceNone = "var(--space-none-xk6qzf, 0px)";
export var spaceXxxs = "var(--space-xxxs-pajhad, 2px)";
export var spaceXxs = "var(--space-xxs-hwfkai, 4px)";
export var spaceXs = "var(--space-xs-ymlm0b, 8px)";
export var spaceS = "var(--space-s-tvghoh, 12px)";
export var spaceM = "var(--space-m-dsumyt, 16px)";
export var spaceL = "var(--space-l-2ud1p3, 20px)";
export var spaceXl = "var(--space-xl-jfy3x4, 24px)";
export var spaceXxl = "var(--space-xxl-32srm4, 32px)";
export var spaceXxxl = "var(--space-xxxl-aut1u7, 40px)";
export var shadowContainer = "var(--shadow-container-53ltfv, 0px 0px 1px 1px #e9ebed, 0px 1px 8px 2px rgba(0, 7, 22, 0.12))";
export var shadowContainerActive = "var(--shadow-container-active-ypjjoc, 0px 1px 1px 1px #e9ebed, 0px 6px 36px #0007161a)";
export var shadowDropdown = "var(--shadow-dropdown-isf0w4, 0px 4px 20px 1px rgba(0, 7, 22, 0.10))";
export var shadowDropup = "var(--shadow-dropup-2r02r5, 0px 4px 20px 1px rgba(0, 7, 22, 0.10))";
export var shadowFlashCollapsed = "var(--shadow-flash-collapsed-b68ip6, 0px 4px 4px rgba(0, 0, 0, 0.25))";
export var shadowFlashSticky = "var(--shadow-flash-sticky-k69vye, 0px 4px 8px rgba(0, 7, 22, 0.10))";
export var shadowModal = "var(--shadow-modal-kwgqht, 0px 4px 20px 1px rgba(0, 7, 22, 0.10))";
export var shadowPanel = "var(--shadow-panel-vk7iea, 0px 0px 0px 1px #b6bec9)";
export var shadowPanelToggle = "var(--shadow-panel-toggle-qddz27, 0px 6px 12px 1px rgba(0, 7, 22, 0.12))";
export var shadowPopover = "var(--shadow-popover-pkane9, 0px 4px 20px 1px rgba(0, 7, 22, 0.10))";
export var shadowSplitBottom = "var(--shadow-split-bottom-vlyulf, 0px -36px 36px -36px rgba(0, 7, 22, 0.10))";
export var shadowSplitSide = "var(--shadow-split-side-nyajix, -1px 0px 1px 0px #e9ebed, -36px 6px 36px -36px rgba(0, 7, 22, 0.10))";
export var shadowSticky = "var(--shadow-sticky-lolw8j, 0px 4px 8px 1px rgba(0, 7, 22, 0.10))";
export var shadowStickyEmbedded = "var(--shadow-sticky-embedded-jmny8n, 0px 2px 0px 0px #e9ebed, 0px 16px 16px -12px rgba(0, 7, 22, 0.10))";
export var shadowStickyColumnFirst = "var(--shadow-sticky-column-first-trcd2o, 4px 0px 8px 1px rgba(0, 7, 22, 0.1))";
export var shadowStickyColumnLast = "var(--shadow-sticky-column-last-qgh697, -4px 0 8px 1px rgba(0, 28, 36, 0.1))";
