// @cloudscape-design/components/test-utils/selectors/prompt-input/index.js
"use strict";
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
const selectors_1 = require("@cloudscape-design/test-utils-core/selectors");
const styles_selectors_js_1 = require("../../../prompt-input/test-classes/styles.selectors.js");
class PromptInputWrapper extends selectors_1.ComponentWrapper {
    findNativeTextarea() {
        return this.findByClassName(styles_selectors_js_1.default.textarea);
    }
    /**
     * Finds the action button. Note that, despite its typings, this may return null.
     */
    findActionButton() {
        return this.findByClassName(styles_selectors_js_1.default['action-button']);
    }
    /**
     * Finds the secondary actions slot. Note that, despite its typings, this may return null.
     */
    findSecondaryActions() {
        return this.findByClassName(styles_selectors_js_1.default['secondary-actions']);
    }
    findSecondaryContent() {
        return this.findByClassName(styles_selectors_js_1.default['secondary-content']);
    }
    findCustomPrimaryAction() {
        return this.findByClassName(styles_selectors_js_1.default['primary-action']);
    }
}
PromptInputWrapper.rootSelector = styles_selectors_js_1.default.root;
exports.default = PromptInputWrapper;
