// @cloudscape-design/components/error-boundary/internal.js
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import React, { Component, createContext, useContext, useState } from 'react';
import { useMergeRefs } from '@cloudscape-design/component-toolkit/internal';
import { ErrorBoundaryFallback } from './fallback';
import styles from './styles.css.js';
const RootSuppressed = Symbol();
const ErrorBoundariesContext = createContext({
    onError: () => { },
    suppressed: RootSuppressed,
});
export function InternalErrorBoundary({ children, suppressNested, suppressible, __internalRootRef, ...props }) {
    const context = useContext(ErrorBoundariesContext);
    const thisSuppressed = context.suppressed === true && suppressible;
    const nextSuppressed = suppressNested || thisSuppressed;
    const [forcedError, setForcedError] = useState(false);
    const setElement = (node) => {
        if (node) {
            if (!node.__awsui__) {
                node.__awsui__ = {};
            }
            node.__awsui__.forceError = () => setForcedError(true);
            node.__awsui__.clearForcedError = () => setForcedError(false);
        }
    };
    const mergedRef = useMergeRefs(setElement, __internalRootRef);
    return (React.createElement("div", { ref: mergedRef, className: styles['error-boundary'], "data-awsui-boundary-id": props.errorBoundaryId }, !thisSuppressed ? (React.createElement(ErrorBoundaryImpl, { ...props, forcedError: forcedError },
        React.createElement(ErrorBoundariesContext.Provider, { value: { ...props, suppressed: nextSuppressed } }, children))) : (React.createElement(React.Fragment, null, children))));
}
export function BuiltInErrorBoundary({ wrapper, suppressNested = false, children }) {
    const context = useContext(ErrorBoundariesContext);
    const thisSuppressed = context.suppressed === true || context.suppressed === RootSuppressed;
    const nextSuppressed = suppressNested || thisSuppressed;
    return !thisSuppressed ? (React.createElement(ErrorBoundaryImpl, { ...context, wrapper: wrapper },
        React.createElement(ErrorBoundariesContext.Provider, { value: { ...context, suppressed: nextSuppressed } }, children))) : (React.createElement(React.Fragment, null, children));
}
class ErrorBoundaryImpl extends Component {
    constructor(props) {
        super(props);
        this.state = { hasError: false };
    }
    static getDerivedStateFromError() {
        return { hasError: true };
    }
    componentDidCatch(error, errorInfo) {
        var _a, _b;
        (_b = (_a = this.props).onError) === null || _b === void 0 ? void 0 : _b.call(_a, { error, errorInfo, errorBoundaryId: this.props.errorBoundaryId });
    }
    render() {
        var _a, _b, _c;
        if (this.state.hasError || this.props.forcedError) {
            const fallback = React.createElement(ErrorBoundaryFallback, { ...this.props });
            return (_c = (_b = (_a = this.props).wrapper) === null || _b === void 0 ? void 0 : _b.call(_a, fallback)) !== null && _c !== void 0 ? _c : fallback;
        }
        return this.props.children;
    }
}
