// @cloudscape-design/components/test-utils/selectors/internal/option.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
const selectors_1 = require("@cloudscape-design/test-utils-core/selectors");
const styles_selectors_js_1 = require("../../../internal/components/option/styles.selectors.js");
const styles_selectors_js_2 = require("../../../select/parts/styles.selectors.js");
class OptionWrapper extends selectors_1.ComponentWrapper {
    /**
     * Finds the label wrapper of this option.
     * @returns {ElementWrapper} the ElementWrapper for the label.
     */
    findLabel() {
        const labelElementWrapper = this.findByClassName(styles_selectors_js_1.default.label);
        if (!labelElementWrapper) {
            // Fallback, if label is null due to custom content.
            return this.findByClassName(styles_selectors_js_1.default['custom-content']);
        }
        return labelElementWrapper;
    }
    findDescription() {
        return this.findByClassName(styles_selectors_js_1.default.description);
    }
    findLabelTag() {
        return this.findByClassName(styles_selectors_js_1.default['label-tag']);
    }
    findTags() {
        return this.findAllByClassName(styles_selectors_js_1.default.tag);
    }
    findDisabledReason() {
        return (0, selectors_1.createWrapper)().find(`.${styles_selectors_js_2.default['disabled-reason-tooltip']}`);
    }
}
OptionWrapper.rootSelector = styles_selectors_js_1.default.option;
exports.default = OptionWrapper;
