// @cloudscape-design/design-tokens/index.js
export var colorChartsRed300 = "var(--color-charts-red-300-2k7eul, #ea7158)";
export var colorChartsRed400 = "var(--color-charts-red-400-ssrf2o, #dc5032)";
export var colorChartsRed500 = "var(--color-charts-red-500-m14kmu, #d13313)";
export var colorChartsRed600 = "var(--color-charts-red-600-938v3h, #ba2e0f)";
export var colorChartsRed700 = "var(--color-charts-red-700-f6sq8t, #a82a0c)";
export var colorChartsRed800 = "var(--color-charts-red-800-tzkaad, #972709)";
export var colorChartsRed900 = "var(--color-charts-red-900-fhg0lh, #892407)";
export var colorChartsRed1000 = "var(--color-charts-red-1000-9iigzo, #7d2105)";
export var colorChartsRed1100 = "var(--color-charts-red-1100-4n7b3z, #721e03)";
export var colorChartsRed1200 = "var(--color-charts-red-1200-ek3cuo, #671c00)";
export var colorChartsOrange300 = "var(--color-charts-orange-300-hqhtmn, #e07941)";
export var colorChartsOrange400 = "var(--color-charts-orange-400-g8c1fc, #cc5f21)";
export var colorChartsOrange500 = "var(--color-charts-orange-500-j3c2cu, #bc4d01)";
export var colorChartsOrange600 = "var(--color-charts-orange-600-1ad7o4, #a84401)";
export var colorChartsOrange700 = "var(--color-charts-orange-700-spsf2r, #983c02)";
export var colorChartsOrange800 = "var(--color-charts-orange-800-244d7b, #8a3603)";
export var colorChartsOrange900 = "var(--color-charts-orange-900-8omk92, #7e3103)";
export var colorChartsOrange1000 = "var(--color-charts-orange-1000-ezq5pz, #732c02)";
export var colorChartsOrange1100 = "var(--color-charts-orange-1100-bhcmg5, #692801)";
export var colorChartsOrange1200 = "var(--color-charts-orange-1200-exs6jj, #602400)";
export var colorChartsYellow300 = "var(--color-charts-yellow-300-fpz8o0, #b2911c)";
export var colorChartsYellow400 = "var(--color-charts-yellow-400-vxiqrf, #9c7b0b)";
export var colorChartsYellow500 = "var(--color-charts-yellow-500-1qgrtj, #8a6b05)";
export var colorChartsYellow600 = "var(--color-charts-yellow-600-aweqy2, #7b5f04)";
export var colorChartsYellow700 = "var(--color-charts-yellow-700-xh0lj9, #6f5504)";
export var colorChartsYellow800 = "var(--color-charts-yellow-800-gp4422, #654d03)";
export var colorChartsYellow900 = "var(--color-charts-yellow-900-r6gx3k, #5d4503)";
export var colorChartsYellow1000 = "var(--color-charts-yellow-1000-6dnac6, #553f03)";
export var colorChartsYellow1100 = "var(--color-charts-yellow-1100-fqp4sw, #4d3901)";
export var colorChartsYellow1200 = "var(--color-charts-yellow-1200-k7kf4w, #483300)";
export var colorChartsGreen300 = "var(--color-charts-green-300-6766ev, #67a353)";
export var colorChartsGreen400 = "var(--color-charts-green-400-gd41ay, #41902c)";
export var colorChartsGreen500 = "var(--color-charts-green-500-yr18n3, #1f8104)";
export var colorChartsGreen600 = "var(--color-charts-green-600-b1gmr0, #1a7302)";
export var colorChartsGreen700 = "var(--color-charts-green-700-305sle, #176702)";
export var colorChartsGreen800 = "var(--color-charts-green-800-rh42zr, #145d02)";
export var colorChartsGreen900 = "var(--color-charts-green-900-2x5smm, #125502)";
export var colorChartsGreen1000 = "var(--color-charts-green-1000-opphoq, #104d01)";
export var colorChartsGreen1100 = "var(--color-charts-green-1100-o67uzm, #0f4601)";
export var colorChartsGreen1200 = "var(--color-charts-green-1200-sw46fc, #0d4000)";
export var colorChartsTeal300 = "var(--color-charts-teal-300-2qlyrg, #2ea597)";
export var colorChartsTeal400 = "var(--color-charts-teal-400-s8pa77, #1c8e81)";
export var colorChartsTeal500 = "var(--color-charts-teal-500-8d830b, #0d7d70)";
export var colorChartsTeal600 = "var(--color-charts-teal-600-772n9t, #096f64)";
export var colorChartsTeal700 = "var(--color-charts-teal-700-29mnwm, #06645a)";
export var colorChartsTeal800 = "var(--color-charts-teal-800-vp41t6, #045b52)";
export var colorChartsTeal900 = "var(--color-charts-teal-900-k020ya, #03524a)";
export var colorChartsTeal1000 = "var(--color-charts-teal-1000-fhpqt3, #014b44)";
export var colorChartsTeal1100 = "var(--color-charts-teal-1100-6w598w, #01443e)";
export var colorChartsTeal1200 = "var(--color-charts-teal-1200-0u78my, #003e38)";
export var colorChartsBlue1300 = "var(--color-charts-blue-1-300-pdza0q, #529ccb)";
export var colorChartsBlue1400 = "var(--color-charts-blue-1-400-ajl038, #3184c2)";
export var colorChartsBlue1500 = "var(--color-charts-blue-1-500-9s8gor, #0273bb)";
export var colorChartsBlue1600 = "var(--color-charts-blue-1-600-7ymb7g, #0166ab)";
export var colorChartsBlue1700 = "var(--color-charts-blue-1-700-5qzras, #015b9d)";
export var colorChartsBlue1800 = "var(--color-charts-blue-1-800-awczh4, #015292)";
export var colorChartsBlue1900 = "var(--color-charts-blue-1-900-6wxwzk, #014a87)";
export var colorChartsBlue11000 = "var(--color-charts-blue-1-1000-00005b, #01437d)";
export var colorChartsBlue11100 = "var(--color-charts-blue-1-1100-8nwfwf, #003c75)";
export var colorChartsBlue11200 = "var(--color-charts-blue-1-1200-v60p8b, #00366d)";
export var colorChartsBlue2300 = "var(--color-charts-blue-2-300-g72slq, #688ae8)";
export var colorChartsBlue2400 = "var(--color-charts-blue-2-400-he538m, #5978e3)";
export var colorChartsBlue2500 = "var(--color-charts-blue-2-500-quctxu, #4066df)";
export var colorChartsBlue2600 = "var(--color-charts-blue-2-600-6qav3j, #3759ce)";
export var colorChartsBlue2700 = "var(--color-charts-blue-2-700-sp7t4m, #314fbf)";
export var colorChartsBlue2800 = "var(--color-charts-blue-2-800-q01umt, #2c46b1)";
export var colorChartsBlue2900 = "var(--color-charts-blue-2-900-gog7z2, #273ea5)";
export var colorChartsBlue21000 = "var(--color-charts-blue-2-1000-c13nf8, #23379b)";
export var colorChartsBlue21100 = "var(--color-charts-blue-2-1100-ddk6eo, #1f3191)";
export var colorChartsBlue21200 = "var(--color-charts-blue-2-1200-gt550t, #1b2b88)";
export var colorChartsPurple300 = "var(--color-charts-purple-300-85q036, #a783e1)";
export var colorChartsPurple400 = "var(--color-charts-purple-400-9axh6r, #9469d6)";
export var colorChartsPurple500 = "var(--color-charts-purple-500-rn2jbl, #8456ce)";
export var colorChartsPurple600 = "var(--color-charts-purple-600-26s4rg, #7749bf)";
export var colorChartsPurple700 = "var(--color-charts-purple-700-tv8cvg, #6b40b2)";
export var colorChartsPurple800 = "var(--color-charts-purple-800-h61qlx, #6237a7)";
export var colorChartsPurple900 = "var(--color-charts-purple-900-am452b, #59309d)";
export var colorChartsPurple1000 = "var(--color-charts-purple-1000-uarqpb, #512994)";
export var colorChartsPurple1100 = "var(--color-charts-purple-1100-y8ctnd, #4a238b)";
export var colorChartsPurple1200 = "var(--color-charts-purple-1200-hr9f40, #431d84)";
export var colorChartsPink300 = "var(--color-charts-pink-300-ewnht7, #da7596)";
export var colorChartsPink400 = "var(--color-charts-pink-400-smjdat, #ce567c)";
export var colorChartsPink500 = "var(--color-charts-pink-500-bw864b, #c33d69)";
export var colorChartsPink600 = "var(--color-charts-pink-600-2ro14y, #b1325c)";
export var colorChartsPink700 = "var(--color-charts-pink-700-ryxvua, #a32952)";
export var colorChartsPink800 = "var(--color-charts-pink-800-tcusf8, #962249)";
export var colorChartsPink900 = "var(--color-charts-pink-900-kpyne4, #8b1b42)";
export var colorChartsPink1000 = "var(--color-charts-pink-1000-1soluc, #81143b)";
export var colorChartsPink1100 = "var(--color-charts-pink-1100-ff6g93, #780d35)";
export var colorChartsPink1200 = "var(--color-charts-pink-1200-w9585d, #6f062f)";
export var colorChartsStatusCritical = "var(--color-charts-status-critical-c6brdu, #7d2105)";
export var colorChartsStatusHigh = "var(--color-charts-status-high-18fhg5, #ba2e0f)";
export var colorChartsStatusMedium = "var(--color-charts-status-medium-3trmy3, #cc5f21)";
export var colorChartsStatusLow = "var(--color-charts-status-low-br6wv0, #b2911c)";
export var colorChartsStatusPositive = "var(--color-charts-status-positive-md7eqa, #67a353)";
export var colorChartsStatusInfo = "var(--color-charts-status-info-yds4x2, #3184c2)";
export var colorChartsStatusNeutral = "var(--color-charts-status-neutral-k2p33t, #8c8c94)";
export var colorChartsThresholdNegative = "var(--color-charts-threshold-negative-aad26m, #db0000)";
export var colorChartsThresholdPositive = "var(--color-charts-threshold-positive-mk2804, #00802f)";
export var colorChartsThresholdInfo = "var(--color-charts-threshold-info-ijuzzj, #006ce0)";
export var colorChartsThresholdNeutral = "var(--color-charts-threshold-neutral-pd7kh4, #656871)";
export var colorChartsLineGrid = "var(--color-charts-line-grid-kjxf3m, #dedee3)";
export var colorChartsLineTick = "var(--color-charts-line-tick-xmcbvk, #dedee3)";
export var colorChartsLineAxis = "var(--color-charts-line-axis-b95ncf, #dedee3)";
export var colorChartsPaletteCategorical1 = "var(--color-charts-palette-categorical-1-xu0deg, #688ae8)";
export var colorChartsPaletteCategorical2 = "var(--color-charts-palette-categorical-2-ktit09, #c33d69)";
export var colorChartsPaletteCategorical3 = "var(--color-charts-palette-categorical-3-g0srj0, #2ea597)";
export var colorChartsPaletteCategorical4 = "var(--color-charts-palette-categorical-4-5vauwp, #8456ce)";
export var colorChartsPaletteCategorical5 = "var(--color-charts-palette-categorical-5-3v8ery, #e07941)";
export var colorChartsPaletteCategorical6 = "var(--color-charts-palette-categorical-6-ztdd8d, #3759ce)";
export var colorChartsPaletteCategorical7 = "var(--color-charts-palette-categorical-7-3j5o6w, #962249)";
export var colorChartsPaletteCategorical8 = "var(--color-charts-palette-categorical-8-c5r39m, #096f64)";
export var colorChartsPaletteCategorical9 = "var(--color-charts-palette-categorical-9-8n6iuv, #6237a7)";
export var colorChartsPaletteCategorical10 = "var(--color-charts-palette-categorical-10-opta0w, #a84401)";
export var colorChartsPaletteCategorical11 = "var(--color-charts-palette-categorical-11-b2r7jc, #273ea5)";
export var colorChartsPaletteCategorical12 = "var(--color-charts-palette-categorical-12-b5drtm, #780d35)";
export var colorChartsPaletteCategorical13 = "var(--color-charts-palette-categorical-13-c69xg9, #03524a)";
export var colorChartsPaletteCategorical14 = "var(--color-charts-palette-categorical-14-db19x8, #4a238b)";
export var colorChartsPaletteCategorical15 = "var(--color-charts-palette-categorical-15-8z8vjw, #7e3103)";
export var colorChartsPaletteCategorical16 = "var(--color-charts-palette-categorical-16-549jkl, #1b2b88)";
export var colorChartsPaletteCategorical17 = "var(--color-charts-palette-categorical-17-nrio7t, #ce567c)";
export var colorChartsPaletteCategorical18 = "var(--color-charts-palette-categorical-18-tm902v, #003e38)";
export var colorChartsPaletteCategorical19 = "var(--color-charts-palette-categorical-19-ujcr86, #9469d6)";
export var colorChartsPaletteCategorical20 = "var(--color-charts-palette-categorical-20-h55e4g, #602400)";
export var colorChartsPaletteCategorical21 = "var(--color-charts-palette-categorical-21-vs0u8l, #4066df)";
export var colorChartsPaletteCategorical22 = "var(--color-charts-palette-categorical-22-6klt3l, #a32952)";
export var colorChartsPaletteCategorical23 = "var(--color-charts-palette-categorical-23-3zpkdt, #0d7d70)";
export var colorChartsPaletteCategorical24 = "var(--color-charts-palette-categorical-24-z9a4uk, #6b40b2)";
export var colorChartsPaletteCategorical25 = "var(--color-charts-palette-categorical-25-tgdsk2, #bc4d01)";
export var colorChartsPaletteCategorical26 = "var(--color-charts-palette-categorical-26-lo8zn9, #2c46b1)";
export var colorChartsPaletteCategorical27 = "var(--color-charts-palette-categorical-27-bruhsa, #81143b)";
export var colorChartsPaletteCategorical28 = "var(--color-charts-palette-categorical-28-6b00fb, #045b52)";
export var colorChartsPaletteCategorical29 = "var(--color-charts-palette-categorical-29-aurmid, #512994)";
export var colorChartsPaletteCategorical30 = "var(--color-charts-palette-categorical-30-fjnmd7, #8a3603)";
export var colorChartsPaletteCategorical31 = "var(--color-charts-palette-categorical-31-7zcct5, #1f3191)";
export var colorChartsPaletteCategorical32 = "var(--color-charts-palette-categorical-32-rrda6y, #da7596)";
export var colorChartsPaletteCategorical33 = "var(--color-charts-palette-categorical-33-2v0mzv, #01443e)";
export var colorChartsPaletteCategorical34 = "var(--color-charts-palette-categorical-34-g9a9q3, #a783e1)";
export var colorChartsPaletteCategorical35 = "var(--color-charts-palette-categorical-35-u0w821, #692801)";
export var colorChartsPaletteCategorical36 = "var(--color-charts-palette-categorical-36-tthuf8, #5978e3)";
export var colorChartsPaletteCategorical37 = "var(--color-charts-palette-categorical-37-y588bl, #b1325c)";
export var colorChartsPaletteCategorical38 = "var(--color-charts-palette-categorical-38-qdh97u, #1c8e81)";
export var colorChartsPaletteCategorical39 = "var(--color-charts-palette-categorical-39-yisq6l, #7749bf)";
export var colorChartsPaletteCategorical40 = "var(--color-charts-palette-categorical-40-yeer1v, #cc5f21)";
export var colorChartsPaletteCategorical41 = "var(--color-charts-palette-categorical-41-tu9dxw, #314fbf)";
export var colorChartsPaletteCategorical42 = "var(--color-charts-palette-categorical-42-q410kp, #8b1b42)";
export var colorChartsPaletteCategorical43 = "var(--color-charts-palette-categorical-43-dwew7q, #06645a)";
export var colorChartsPaletteCategorical44 = "var(--color-charts-palette-categorical-44-2thp96, #59309d)";
export var colorChartsPaletteCategorical45 = "var(--color-charts-palette-categorical-45-6kinj6, #983c02)";
export var colorChartsPaletteCategorical46 = "var(--color-charts-palette-categorical-46-iefxfq, #23379b)";
export var colorChartsPaletteCategorical47 = "var(--color-charts-palette-categorical-47-9l9wl2, #6f062f)";
export var colorChartsPaletteCategorical48 = "var(--color-charts-palette-categorical-48-5s2n0r, #014b44)";
export var colorChartsPaletteCategorical49 = "var(--color-charts-palette-categorical-49-bub0l1, #431d84)";
export var colorChartsPaletteCategorical50 = "var(--color-charts-palette-categorical-50-utrpu5, #732c02)";
export var colorChartsErrorBarMarker = "var(--color-charts-error-bar-marker-r10jgv, #131920)";
export var colorBackgroundNotificationSeverityCritical = "var(--color-background-notification-severity-critical-0xl8pp, #870303)";
export var colorBackgroundNotificationSeverityHigh = "var(--color-background-notification-severity-high-8nbgdi, #ce3311)";
export var colorBackgroundNotificationSeverityMedium = "var(--color-background-notification-severity-medium-lbljs2, #f89256)";
export var colorBackgroundNotificationSeverityLow = "var(--color-background-notification-severity-low-giz8b6, #f2cd54)";
export var colorBackgroundNotificationSeverityNeutral = "var(--color-background-notification-severity-neutral-hnhgmv, #656871)";
export var colorTextNotificationSeverityCritical = "var(--color-text-notification-severity-critical-tv4vw4, #f9f9fa)";
export var colorTextNotificationSeverityHigh = "var(--color-text-notification-severity-high-t4suvu, #f9f9fa)";
export var colorTextNotificationSeverityMedium = "var(--color-text-notification-severity-medium-8f60kb, #0f141a)";
export var colorTextNotificationSeverityLow = "var(--color-text-notification-severity-low-gvojhi, #0f141a)";
export var colorTextNotificationSeverityNeutral = "var(--color-text-notification-severity-neutral-ynm2wl, #f9f9fa)";
export var colorBackgroundButtonLinkActive = "var(--color-background-button-link-active-5oi2dp, #d1f1ff)";
export var colorBackgroundButtonLinkHover = "var(--color-background-button-link-hover-lhrs2u, #f0fbff)";
export var colorBackgroundButtonNormalActive = "var(--color-background-button-normal-active-5imwxd, #d1f1ff)";
export var colorBackgroundButtonNormalDefault = "var(--color-background-button-normal-default-7f99mv, #ffffff)";
export var colorBackgroundButtonNormalDisabled = "var(--color-background-button-normal-disabled-hl039l, #ffffff)";
export var colorBackgroundButtonNormalHover = "var(--color-background-button-normal-hover-53op9s, #f0fbff)";
export var colorBackgroundToggleButtonNormalPressed = "var(--color-background-toggle-button-normal-pressed-4khex7, #d1f1ff)";
export var colorBackgroundButtonPrimaryActive = "var(--color-background-button-primary-active-5cqoqt, #002b66)";
export var colorBackgroundButtonPrimaryDefault = "var(--color-background-button-primary-default-vdt0fu, #006ce0)";
export var colorBackgroundButtonPrimaryDisabled = "var(--color-background-button-primary-disabled-sgo4zo, #ebebf0)";
export var colorBackgroundButtonPrimaryHover = "var(--color-background-button-primary-hover-mo85i6, #002b66)";
export var colorBackgroundCellShaded = "var(--color-background-cell-shaded-v7o6so, #f6f6f9)";
export var colorBackgroundContainerContent = "var(--color-background-container-content-6u8rvp, #ffffff)";
export var colorBackgroundContainerHeader = "var(--color-background-container-header-gs3mbe, #ffffff)";
export var colorBackgroundControlChecked = "var(--color-background-control-checked-ka7kc2, #006ce0)";
export var colorBackgroundControlDefault = "var(--color-background-control-default-4jb21l, #ffffff)";
export var colorBackgroundControlDisabled = "var(--color-background-control-disabled-1f3718, #dedee3)";
export var colorBackgroundDropdownItemDefault = "var(--color-background-dropdown-item-default-qmc033, #ffffff)";
export var colorBackgroundDropdownItemFilterMatch = "var(--color-background-dropdown-item-filter-match-49b5vt, #f0fbff)";
export var colorBackgroundDropdownItemHover = "var(--color-background-dropdown-item-hover-yunepc, #f3f3f7)";
export var colorBackgroundHomeHeader = "var(--color-background-home-header-4c9jt4, #0f141a)";
export var colorBackgroundInputDefault = "var(--color-background-input-default-ifz5bb, #ffffff)";
export var colorBackgroundInputDisabled = "var(--color-background-input-disabled-dihaja, #ebebf0)";
export var colorBackgroundItemSelected = "var(--color-background-item-selected-9gppru, #f0fbff)";
export var colorBackgroundLayoutMain = "var(--color-background-layout-main-5ilwcb, #ffffff)";
export var colorBackgroundLayoutToggleActive = "var(--color-background-layout-toggle-active-ap91vm, #424650)";
export var colorBackgroundLayoutToggleDefault = "var(--color-background-layout-toggle-default-2hgjdu, #424650)";
export var colorBackgroundLayoutToggleHover = "var(--color-background-layout-toggle-hover-0cpm7g, #656871)";
export var colorBackgroundLayoutToggleSelectedActive = "var(--color-background-layout-toggle-selected-active-zcl8w3, #006ce0)";
export var colorBackgroundLayoutToggleSelectedDefault = "var(--color-background-layout-toggle-selected-default-izfana, #006ce0)";
export var colorBackgroundLayoutToggleSelectedHover = "var(--color-background-layout-toggle-selected-hover-7953u1, #004a9e)";
export var colorBackgroundNotificationBlue = "var(--color-background-notification-blue-4vnob8, #006ce0)";
export var colorBackgroundNotificationGreen = "var(--color-background-notification-green-2rkyvu, #00802f)";
export var colorBackgroundNotificationGrey = "var(--color-background-notification-grey-x3vul6, #424650)";
export var colorBackgroundNotificationRed = "var(--color-background-notification-red-0487ea, #db0000)";
export var colorBackgroundNotificationYellow = "var(--color-background-notification-yellow-y6us5r, #ffe347)";
export var colorBackgroundPopover = "var(--color-background-popover-e20fy8, #ffffff)";
export var colorBackgroundProgressBarValueDefault = "var(--color-background-progress-bar-value-default-69ydqg, #006ce0)";
export var colorBackgroundProgressBarDefault = "var(--color-background-progress-bar-default-j8kyxd, #ebebf0)";
export var colorBackgroundSegmentActive = "var(--color-background-segment-active-1u2ldl, #006ce0)";
export var colorBackgroundSegmentDefault = "var(--color-background-segment-default-b0r494, #ffffff)";
export var colorBackgroundSegmentDisabled = "var(--color-background-segment-disabled-m2a5t7, #ffffff)";
export var colorBackgroundSegmentHover = "var(--color-background-segment-hover-800sl4, #f0fbff)";
export var colorBackgroundSliderRangeDefault = "var(--color-background-slider-range-default-3ljdu4, #006ce0)";
export var colorBackgroundSliderRangeActive = "var(--color-background-slider-range-active-vu3lky, #004a9e)";
export var colorBackgroundSliderHandleDefault = "var(--color-background-slider-handle-default-lp5ntg, #006ce0)";
export var colorBackgroundSliderHandleActive = "var(--color-background-slider-handle-active-50ubqb, #004a9e)";
export var colorBackgroundSliderTrackDefault = "var(--color-background-slider-track-default-vk2c9o, #8c8c94)";
export var colorBackgroundStatusError = "var(--color-background-status-error-mu3lcw, #fff5f5)";
export var colorBackgroundStatusInfo = "var(--color-background-status-info-sfobba, #f0fbff)";
export var colorBackgroundDialog = "var(--color-background-dialog-2fj3uu, #f0fbff)";
export var colorBackgroundStatusSuccess = "var(--color-background-status-success-h6b8bh, #effff1)";
export var colorBackgroundStatusWarning = "var(--color-background-status-warning-cv83up, #fffef0)";
export var colorBackgroundToggleCheckedDisabled = "var(--color-background-toggle-checked-disabled-amxc0a, #b8e7ff)";
export var colorBackgroundAvatarGenAi = "var(--color-background-avatar-gen-ai-oxp2v6, radial-gradient(circle farthest-corner at top right, #b8e7ff 0%, #0099ff 25%, #5c7fff 40% , #8575ff 60%, #962eff 80%))";
export var colorBackgroundAvatarDefault = "var(--color-background-avatar-default-t427xm, #424650)";
export var colorTextAvatar = "var(--color-text-avatar-kuhkoa, #ffffff)";
export var colorBackgroundLoadingBarGenAi = "var(--color-background-loading-bar-gen-ai-tey70i, linear-gradient(90deg, #b8e7ff 0%, #0099ff 10%, #5c7fff 24%, #8575ff 50%, #962eff 76%, #0099ff 90%, #b8e7ff 100%))";
export var colorBackgroundChatBubbleOutgoing = "var(--color-background-chat-bubble-outgoing-ay6nj3, transparent)";
export var colorBackgroundChatBubbleIncoming = "var(--color-background-chat-bubble-incoming-j38cew, #f6f6f9)";
export var colorTextChatBubbleOutgoing = "var(--color-text-chat-bubble-outgoing-f3r63s, #0f141a)";
export var colorTextChatBubbleIncoming = "var(--color-text-chat-bubble-incoming-od0yh8, #0f141a)";
export var colorBorderButtonNormalActive = "var(--color-border-button-normal-active-ru7yhb, #002b66)";
export var colorBorderButtonNormalDefault = "var(--color-border-button-normal-default-glqfp1, #006ce0)";
export var colorBorderToggleButtonNormalPressed = "var(--color-border-toggle-button-normal-pressed-tq8o41, #006ce0)";
export var colorBorderButtonNormalDisabled = "var(--color-border-button-normal-disabled-pkhetz, #b4b4bb)";
export var colorTextButtonNormalDisabled = "var(--color-text-button-normal-disabled-05p74s, #8c8c94)";
export var colorBorderButtonNormalHover = "var(--color-border-button-normal-hover-6a2tdq, #002b66)";
export var colorTextButtonIconDisabled = "var(--color-text-button-icon-disabled-nnofkn, #8c8c94)";
export var colorBorderButtonPrimaryDisabled = "var(--color-border-button-primary-disabled-b5p1ji, #ebebf0)";
export var colorTextButtonPrimaryDisabled = "var(--color-text-button-primary-disabled-q79gms, #8c8c94)";
export var colorItemSelected = "var(--color-item-selected-72rnwy, #006ce0)";
export var colorBorderContainerTop = "var(--color-border-container-top-k3vmoz, transparent)";
export var colorBorderControlDefault = "var(--color-border-control-default-sh3548, #8c8c94)";
export var colorBorderDividerDefault = "var(--color-border-divider-default-nr68jt, #c6c6cd)";
export var colorBorderDividerSecondary = "var(--color-border-divider-secondary-qoitch, #ebebf0)";
export var colorBorderDropdownContainer = "var(--color-border-dropdown-container-cmthq7, #b4b4bb)";
export var colorBorderDropdownItemHover = "var(--color-border-dropdown-item-hover-aqfuxq, #8c8c94)";
export var colorBorderInputDefault = "var(--color-border-input-default-317xk5, #8c8c94)";
export var colorBorderInputFocused = "var(--color-border-input-focused-4z0pgn, #006ce0)";
export var colorBorderItemFocused = "var(--color-border-item-focused-uk47pl, #006ce0)";
export var colorBorderDropdownItemFocused = "var(--color-border-dropdown-item-focused-zacqlp, #424650)";
export var colorBorderItemSelected = "var(--color-border-item-selected-wl5ttm, #006ce0)";
export var colorBorderPopover = "var(--color-border-popover-1ye6tz, #b4b4bb)";
export var colorBorderSegmentActive = "var(--color-border-segment-active-ls9t4n, #424650)";
export var colorBorderSegmentDefault = "var(--color-border-segment-default-6ig2mo, #424650)";
export var colorBorderSegmentDisabled = "var(--color-border-segment-disabled-fcrbcl, #424650)";
export var colorBorderSegmentHover = "var(--color-border-segment-hover-0o9ey3, #424650)";
export var colorBorderStatusError = "var(--color-border-status-error-j8acpp, #db0000)";
export var colorBorderStatusInfo = "var(--color-border-status-info-qf6jok, #006ce0)";
export var colorBorderStatusSuccess = "var(--color-border-status-success-8z5f8u, #00802f)";
export var colorBorderStatusWarning = "var(--color-border-status-warning-j40pg7, #855900)";
export var colorBorderDialog = "var(--color-border-dialog-0rjwug, #006ce0)";
export var colorForegroundControlDefault = "var(--color-foreground-control-default-eto4wy, #ffffff)";
export var colorForegroundControlDisabled = "var(--color-foreground-control-disabled-txi6cf, #ffffff)";
export var colorForegroundControlReadOnly = "var(--color-foreground-control-read-only-7ydvuj, #656871)";
export var colorTextAccent = "var(--color-text-accent-n1kmht, #006ce0)";
export var colorTextBodyDefault = "var(--color-text-body-default-vvtq8u, #0f141a)";
export var colorTextBodySecondary = "var(--color-text-body-secondary-yna5sb, #424650)";
export var colorTextBreadcrumbCurrent = "var(--color-text-breadcrumb-current-2mqnkk, #656871)";
export var colorTextBreadcrumbIcon = "var(--color-text-breadcrumb-icon-9j48ot, #8c8c94)";
export var colorTextButtonNormalActive = "var(--color-text-button-normal-active-vihsxh, #002b66)";
export var colorTextToggleButtonNormalPressed = "var(--color-text-toggle-button-normal-pressed-wnx2zl, #002b66)";
export var colorTextButtonNormalDefault = "var(--color-text-button-normal-default-nzalii, #006ce0)";
export var colorTextButtonNormalHover = "var(--color-text-button-normal-hover-gusgyv, #002b66)";
export var colorTextButtonPrimaryActive = "var(--color-text-button-primary-active-refmba, #ffffff)";
export var colorTextButtonPrimaryDefault = "var(--color-text-button-primary-default-mwl31m, #ffffff)";
export var colorTextButtonPrimaryHover = "var(--color-text-button-primary-hover-pw12ep, #ffffff)";
export var colorTextCounter = "var(--color-text-counter-bywf75, #656871)";
export var colorTextDropdownItemDefault = "var(--color-text-dropdown-item-default-f1jr9u, #0f141a)";
export var colorTextDropdownItemDisabled = "var(--color-text-dropdown-item-disabled-8m65hf, #b4b4bb)";
export var colorTextDropdownItemFilterMatch = "var(--color-text-dropdown-item-filter-match-ebhvct, #006ce0)";
export var colorTextDropdownItemHighlighted = "var(--color-text-dropdown-item-highlighted-yr1px8, #0f141a)";
export var colorTextEmpty = "var(--color-text-empty-tlohug, #656871)";
export var colorTextFormDefault = "var(--color-text-form-default-47mtz6, #0f141a)";
export var colorTextFormSecondary = "var(--color-text-form-secondary-1nm780, #656871)";
export var colorTextGroupLabel = "var(--color-text-group-label-a2qc05, #424650)";
export var colorTextLabelGenAi = "var(--color-text-label-gen-ai-a2n3od, #7300e5)";
export var colorTextHeadingDefault = "var(--color-text-heading-default-izpp46, #0f141a)";
export var colorTextHeadingSecondary = "var(--color-text-heading-secondary-iwtvf6, #424650)";
export var colorTextHomeHeaderDefault = "var(--color-text-home-header-default-morg6i, #ebebf0)";
export var colorTextHomeHeaderSecondary = "var(--color-text-home-header-secondary-i4jhp7, #c6c6cd)";
export var colorTextInputDisabled = "var(--color-text-input-disabled-wh1f3y, #b4b4bb)";
export var colorTextInputPlaceholder = "var(--color-text-input-placeholder-dclg8u, #656871)";
export var colorTextInteractiveActive = "var(--color-text-interactive-active-uoe6zi, #0f141a)";
export var colorTextInteractiveDefault = "var(--color-text-interactive-default-ugh9wp, #424650)";
export var colorTextInteractiveDisabled = "var(--color-text-interactive-disabled-1bqmrl, #b4b4bb)";
export var colorTextInteractiveHover = "var(--color-text-interactive-hover-6naf7i, #0f141a)";
export var colorTextToggleButtonIconPressed = "var(--color-text-toggle-button-icon-pressed-detfkz, #0f141a)";
export var colorTextInteractiveInvertedDefault = "var(--color-text-interactive-inverted-default-xlc0d5, #dedee3)";
export var colorTextInteractiveInvertedHover = "var(--color-text-interactive-inverted-hover-65rnp7, #f9f9fa)";
export var colorTextLabel = "var(--color-text-label-28gfmc, #0f141a)";
export var colorTextLayoutToggle = "var(--color-text-layout-toggle-1a15s3, #ffffff)";
export var colorTextLayoutToggleActive = "var(--color-text-layout-toggle-active-ifu7qp, #ffffff)";
export var colorTextLayoutToggleHover = "var(--color-text-layout-toggle-hover-9jwdce, #006ce0)";
export var colorTextLayoutToggleSelected = "var(--color-text-layout-toggle-selected-xpximc, #ffffff)";
export var colorTextLinkDefault = "var(--color-text-link-default-hude44, #006ce0)";
export var colorTextLinkHover = "var(--color-text-link-hover-2hfec2, #002b66)";
export var colorTextNotificationDefault = "var(--color-text-notification-default-1iey72, #f9f9fa)";
export var colorTextSegmentActive = "var(--color-text-segment-active-hlorbe, #ffffff)";
export var colorTextSegmentDefault = "var(--color-text-segment-default-vi2vn9, #424650)";
export var colorTextSegmentHover = "var(--color-text-segment-hover-65a2x8, #002b66)";
export var colorTextStatusError = "var(--color-text-status-error-ksqavh, #db0000)";
export var colorTextStatusInactive = "var(--color-text-status-inactive-gy7337, #656871)";
export var colorTextStatusInfo = "var(--color-text-status-info-ue8bd2, #006ce0)";
export var colorTextStatusSuccess = "var(--color-text-status-success-ybmii8, #00802f)";
export var colorTextStatusWarning = "var(--color-text-status-warning-6meo06, #855900)";
export var colorTextTopNavigationTitle = "var(--color-text-top-navigation-title-en0v40, #0f141a)";
export var colorBoardPlaceholderActive = "var(--color-board-placeholder-active-x6yfem, #ebebf0)";
export var colorBoardPlaceholderHover = "var(--color-board-placeholder-hover-5nov9c, #d1f1ff)";
export var colorDragPlaceholderActive = "var(--color-drag-placeholder-active-ea1sgp, #ebebf0)";
export var colorDragPlaceholderHover = "var(--color-drag-placeholder-hover-qwadna, #d1f1ff)";
export var colorDropzoneBackgroundDefault = "var(--color-dropzone-background-default-7efhmb, #ffffff)";
export var colorDropzoneBackgroundHover = "var(--color-dropzone-background-hover-mi8rlm, #f0fbff)";
export var colorDropzoneTextDefault = "var(--color-dropzone-text-default-djdnme, #424650)";
export var colorDropzoneTextHover = "var(--color-dropzone-text-hover-asw3rt, #424650)";
export var colorDropzoneBorderDefault = "var(--color-dropzone-border-default-k648ha, #8c8c94)";
export var colorDropzoneBorderHover = "var(--color-dropzone-border-hover-otpag5, #002b66)";
export var colorTreeViewConnectorLine = "var(--color-tree-view-connector-line-1usxvn, #8c8c94)";
export var fontFamilyBase = "var(--font-family-base-gmnpzl, 'Open Sans', 'Helvetica Neue', Roboto, Arial, sans-serif)";
export var fontFamilyMonospace = "var(--font-family-monospace-q47m7k, Monaco, Menlo, Consolas, 'Courier Prime', Courier, 'Courier New', monospace)";
export var fontSizeBodyM = "var(--font-size-body-m-a7nh2n, 14px)";
export var fontSizeBodyS = "var(--font-size-body-s-smc8cv, 12px)";
export var fontSizeDisplayL = "var(--font-size-display-l-wa6woo, 42px)";
export var fontSizeHeadingXl = "var(--font-size-heading-xl-wvkbur, 24px)";
export var fontSizeHeadingL = "var(--font-size-heading-l-vnacx6, 20px)";
export var fontSizeHeadingM = "var(--font-size-heading-m-170yiy, 18px)";
export var fontSizeHeadingS = "var(--font-size-heading-s-zp08en, 16px)";
export var fontSizeHeadingXs = "var(--font-size-heading-xs-j8yzxv, 14px)";
export var fontWeightButton = "var(--font-weight-button-0eg20c, 700)";
export var fontWeightHeadingXl = "var(--font-weight-heading-xl-u3m4we, 700)";
export var fontWeightHeadingL = "var(--font-weight-heading-l-0t6dwc, 700)";
export var fontWeightHeadingM = "var(--font-weight-heading-m-zf82dr, 700)";
export var fontWeightHeadingS = "var(--font-weight-heading-s-lcx0ai, 700)";
export var fontWeightHeadingXs = "var(--font-weight-heading-xs-wqqpne, 700)";
export var lineHeightBodyM = "var(--line-height-body-m-2mh3ke, 20px)";
export var lineHeightBodyS = "var(--line-height-body-s-nu5hx1, 16px)";
export var lineHeightDisplayL = "var(--line-height-display-l-vwanzp, 48px)";
export var lineHeightHeadingXl = "var(--line-height-heading-xl-hko6p0, 30px)";
export var lineHeightHeadingL = "var(--line-height-heading-l-mg5bx6, 24px)";
export var lineHeightHeadingM = "var(--line-height-heading-m-uoaqdh, 22px)";
export var lineHeightHeadingS = "var(--line-height-heading-s-hmi4vc, 20px)";
export var lineHeightHeadingXs = "var(--line-height-heading-xs-q9j004, 18px)";
export var borderRadiusAlert = "var(--border-radius-alert-syagf6, 12px)";
export var borderRadiusBadge = "var(--border-radius-badge-exolfb, 4px)";
export var borderRadiusButton = "var(--border-radius-button-7bgkcs, 20px)";
export var borderRadiusCalendarDayFocusRing = "var(--border-radius-calendar-day-focus-ring-xvvbuc, 3px)";
export var borderRadiusContainer = "var(--border-radius-container-nsfwmm, 16px)";
export var borderRadiusControlCircularFocusRing = "var(--border-radius-control-circular-focus-ring-yjhscw, 4px)";
export var borderRadiusControlDefaultFocusRing = "var(--border-radius-control-default-focus-ring-1uabki, 4px)";
export var borderRadiusDropdown = "var(--border-radius-dropdown-fgc2a1, 8px)";
export var borderRadiusDropzone = "var(--border-radius-dropzone-eklq14, 12px)";
export var borderRadiusFlashbar = "var(--border-radius-flashbar-pp1ptu, 12px)";
export var borderRadiusItem = "var(--border-radius-item-iwaia5, 8px)";
export var borderRadiusInput = "var(--border-radius-input-7q0str, 8px)";
export var borderRadiusPopover = "var(--border-radius-popover-6fqb5w, 8px)";
export var borderRadiusTabsFocusRing = "var(--border-radius-tabs-focus-ring-o4qku1, 20px)";
export var borderRadiusTiles = "var(--border-radius-tiles-wm1vgw, 8px)";
export var borderRadiusToken = "var(--border-radius-token-ycnemh, 8px)";
export var borderRadiusChatBubble = "var(--border-radius-chat-bubble-haafsg, 8px)";
export var borderRadiusTutorialPanelItem = "var(--border-radius-tutorial-panel-item-ojaqxg, 8px)";
export var borderWidthAlert = "var(--border-width-alert-tuifgy, 2px)";
export var borderWidthButton = "var(--border-width-button-jm0qg7, 2px)";
export var borderWidthDropdown = "var(--border-width-dropdown-youcay, 2px)";
export var borderWidthField = "var(--border-width-field-2xc78x, 1px)";
export var borderWidthPopover = "var(--border-width-popover-nflirh, 2px)";
export var borderWidthToken = "var(--border-width-token-2ukdpu, 2px)";
export var borderWidthIconSmall = "var(--border-width-icon-small-z55i5t, 2px)";
export var borderWidthIconNormal = "var(--border-width-icon-normal-9h7vj7, 2px)";
export var borderWidthIconMedium = "var(--border-width-icon-medium-b7icqv, 2px)";
export var borderWidthIconBig = "var(--border-width-icon-big-ymgy42, 3px)";
export var borderWidthIconLarge = "var(--border-width-icon-large-u645rg, 4px)";
export var motionDurationAvatarGenAiGradient = "var(--motion-duration-avatar-gen-ai-gradient-84si5n, 3600ms)";
export var motionDurationAvatarLoadingDots = "var(--motion-duration-avatar-loading-dots-1xxvis, 1200ms)";
export var motionEasingAvatarGenAiGradient = "var(--motion-easing-avatar-gen-ai-gradient-9fwaak, cubic-bezier(0.7, 0, 0.3, 1))";
export var motionEasingResponsive = "var(--motion-easing-responsive-hjj3ai, cubic-bezier(0, 0, 0, 1))";
export var motionEasingSticky = "var(--motion-easing-sticky-tn072u, cubic-bezier(1, 0, 0.83, 1))";
export var motionEasingExpressive = "var(--motion-easing-expressive-o5jqzg, cubic-bezier(0.84, 0, 0.16, 1))";
export var motionDurationResponsive = "var(--motion-duration-responsive-mehora, 115ms)";
export var motionDurationExpressive = "var(--motion-duration-expressive-cbdcwy, 165ms)";
export var motionDurationComplex = "var(--motion-duration-complex-tbdo30, 250ms)";
export var motionKeyframesFadeIn = "var(--motion-keyframes-fade-in-0r842q, awsui-fade-in-35003c)";
export var motionKeyframesFadeOut = "var(--motion-keyframes-fade-out-g7fgdu, awsui-fade-out-35003c)";
export var motionKeyframesStatusIconError = "var(--motion-keyframes-status-icon-error-wkou39, awsui-status-icon-error-35003c)";
export var motionKeyframesScalePopup = "var(--motion-keyframes-scale-popup-9iqcu0, awsui-scale-popup-35003c)";
export var spaceContainerHorizontal = "var(--space-container-horizontal-nqrzyh, 20px)";
export var spaceFieldHorizontal = "var(--space-field-horizontal-0aq2ch, 12px)";
export var spaceTreeViewIndentation = "var(--space-tree-view-indentation-xh9kis, 24px)";
export var spaceScaledXxxs = "var(--space-scaled-xxxs-oo06c7, 2px)";
export var spaceScaledXxs = "var(--space-scaled-xxs-pfm1nx, 4px)";
export var spaceScaledXs = "var(--space-scaled-xs-xwoogq, 8px)";
export var spaceScaledS = "var(--space-scaled-s-8ozaad, 12px)";
export var spaceScaledM = "var(--space-scaled-m-m892r9, 16px)";
export var spaceScaledL = "var(--space-scaled-l-sej05l, 20px)";
export var spaceScaledXl = "var(--space-scaled-xl-dunxp5, 24px)";
export var spaceScaledXxl = "var(--space-scaled-xxl-6wgq96, 32px)";
export var spaceScaledXxxl = "var(--space-scaled-xxxl-hwoy7j, 40px)";
export var spaceStaticXxxs = "var(--space-static-xxxs-yidks1, 2px)";
export var spaceStaticXxs = "var(--space-static-xxs-ns94dp, 4px)";
export var spaceStaticXs = "var(--space-static-xs-gnm0mz, 8px)";
export var spaceStaticS = "var(--space-static-s-t763lu, 12px)";
export var spaceStaticM = "var(--space-static-m-m6qboo, 16px)";
export var spaceStaticL = "var(--space-static-l-n53k41, 20px)";
export var spaceStaticXl = "var(--space-static-xl-4tedi6, 24px)";
export var spaceStaticXxl = "var(--space-static-xxl-ifa9j8, 32px)";
export var spaceStaticXxxl = "var(--space-static-xxxl-tngnnz, 40px)";
export var shadowContainerActive = "var(--shadow-container-active-ypjjoc, 0px 1px 1px 1px #e9ebed, 0px 6px 36px #0007161a)";
