// @cloudscape-design/components/test-utils/selectors/token/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
const selectors_1 = require("@cloudscape-design/test-utils-core/selectors");
const option_1 = require("../internal/option");
const styles_selectors_js_1 = require("../../../token/test-classes/styles.selectors.js");
class TokenWrapper extends selectors_1.ComponentWrapper {
    findOption() {
        return this.findComponent(`.${option_1.default.rootSelector}`, option_1.default);
    }
    /**
     * Returns the token label.
     */
    findLabel() {
        return this.findOption().findLabel();
    }
    /**
     * Returns the token label tag.
     */
    findLabelTag() {
        return this.findOption().findLabelTag();
    }
    /**
     * Returns the token description.
     */
    findDescription() {
        return this.findOption().findDescription();
    }
    /**
     * Returns the token tags.
     */
    findTags() {
        return this.findOption().findTags();
    }
    /**
     * Returns the token dismiss button.
     */
    findDismiss() {
        return this.findByClassName(styles_selectors_js_1.default['dismiss-button']);
    }
}
TokenWrapper.rootSelector = styles_selectors_js_1.default.root;
exports.default = TokenWrapper;
